{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8018bc56",
   "metadata": {},
   "source": [
    "<img src=\"30S.png\" width=\"50%\" alt=\"Image of 30S ribosomal subunit\" style=\"float:right\"/>\n",
    "\n",
    "# Ribosomal Database Project\n",
    "\n",
    "The Ribosomal Database Project (RDP) aims to provide a comprehensive database of ribosomal RNA (rRNA) sequences and related data for use in phylogenetic analysis and microbial ecology research. The project was initiated in 1984 by Carl Woese, a pioneer in the field of molecular evolution, and is currently maintained by a team of researchers at Michigan State University.\n",
    "\n",
    "The RDP contains a curated database of rRNA sequences from bacteria, archaea, and eukaryotes. The database includes both 16S and 18S rRNA sequences, which are commonly used as molecular markers for bacterial and eukaryotic phylogenetic analysis, respectively. \n",
    "\n",
    "Many bacteria have more than one copy of 16s rRNA genes in their genome, with some having as many as 15 different copies (doi: 10.1371/journal.pone.0057923. Epub 2013 Feb 27. PMID: 23460914)\n",
    "\n",
    "In this workshop, we'll use a curated file of approximately 15000 16S rDNA sequences."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da276969",
   "metadata": {},
   "source": [
    "# Retrieving your Sequences\n",
    "\n",
    "The 16S sequences obtained by Nanopore sequencing have been provided in a csv file called `class_16S_sequences.csv`.\n",
    "\n",
    "Samples were identified by using extra sequences known as 'barcodes' that were added to the PCR primers. Each sample was tagged with a unique combination from 8 forward and 15 reverse primer sequences - you noted these down in the lab in week 5 when you set up the PCR. Look up your 2 barcode references, for example A2 or F6. \n",
    "\n",
    "# Workshop tasks\n",
    "\n",
    "Today's workshop is divided into 5 sections:\n",
    "* Retrieving your 16S sequences\n",
    "* Loading the reference 16S sequence database\n",
    "* Searching for database matches\n",
    "* Visualising your analysis results\n",
    "* Evaluation and Reflection on Results\n",
    "\n",
    "### Task 1: Retrieving your 16S sequences\n",
    "\n",
    "1.1> Load the file 'Discovery_16S_consensus.csv' into a Pandas dataframe and explore the contents. n.b. the 'descriptor' is a pseudo-random identifier for your sequence (your can ignore it). \n",
    "\n",
    "1.2> Retrieve your 16S sequences from the Pandas dataframe using the `.query()` function by looking for rows that match your combination of forward and reverse primers. If there is no match for your group's 2 pairs of forward and reverse primers, run the code below to show which combinations of primers do have reads, then and and select one you like the look of for further analysis in this workshop. \n",
    "```python \n",
    "#load the dataframe\n",
    "df=pd.read_csv(\"class_16S_sequences.csv\")\n",
    "# Create a contingency table of the forward and reverse indices\n",
    "ct = pd.crosstab(df['reverse_primer'], df['forward_primer'])\n",
    "print(ct)\n",
    "#Add code below this that uses .query to filter the results for your combination of forward and reverse primers. \n",
    "```\n",
    "\n",
    "1.3> Extract your 16S sequences to a list using `.tolist()`. From the list, extracting individual sequences using indexing (for example `my_sequences[0]` or `my_sequences[1]`).\n",
    "\n",
    "```python\n",
    "seq_list=my_sequences['consensus_sequence'].tolist()\n",
    "seq_list[0]\n",
    "```\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6a27ce64",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "forward_primer  A  B  C  D  E  F  G  H  I  J  K  L  U  V  Y\n",
      "reverse_primer                                             \n",
      "0               1  0  3  1  0  0  0  0  1  0  1  1  0  0  0\n",
      "1               0  0  0  0  1  1  0  0  0  0  0  0  0  0  0\n",
      "2               0  1  0  0  1  0  0  0  2  1  0  0  0  0  0\n",
      "3               0  0  1  0  0  0  0  1  0  1  0  1  0  0  0\n",
      "4               0  0  0  0  0  1  0  1  0  0  0  1  0  0  0\n",
      "5               0  0  1  0  0  0  0  0  0  0  0  2  0  0  0\n",
      "6               1  0  0  0  0  1  0  0  0  0  0  0  0  0  0\n",
      "7               0  1  0  0  0  1  0  0  1  0  0  0  0  0  1\n",
      "8               0  1  0  0  0  0  0  0  0  0  0  0  0  0  0\n",
      "9               1  0  0  0  2  0  0  0  2  2  0  0  0  1  0\n",
      "10              0  1  0  0  0  0  0  0  1  0  0  1  0  1  0\n",
      "11              0  0  1  0  0  1  0  0  0  0  0  1  0  0  0\n",
      "12              0  1  1  0  0  0  0  0  0  0  0  0  1  0  0\n",
      "13              0  1  0  2  0  0  2  3  0  0  1  0  0  0  0\n",
      "14              1  2  2  1  0  0  0  0  2  1  1  0  0  0  0\n",
      "15              0  0  0  0  0  0  1  0  0  3  0  0  0  0  0\n",
      "17              0  0  0  0  0  0  1  1  0  0  0  0  0  0  0\n",
      "18              0  0  0  0  0  0  0  1  0  0  0  0  0  0  0\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df=pd.read_csv(\"class_16S_sequences.csv\")\n",
    "ct = pd.crosstab(df['reverse_primer'], df['forward_primer'])\n",
    "print(ct)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "66b32a82",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "forward_primer  A  B  C  D  E  F  G  H  I  J  K  L  U  V  Y\n",
      "reverse_primer                                             \n",
      "0               1  0  3  1  0  0  0  0  1  0  1  1  0  0  0\n",
      "1               0  0  0  0  1  1  0  0  0  0  0  0  0  0  0\n",
      "2               0  1  0  0  1  0  0  0  2  1  0  0  0  0  0\n",
      "3               0  0  1  0  0  0  0  1  0  1  0  1  0  0  0\n",
      "4               0  0  0  0  0  1  0  1  0  0  0  1  0  0  0\n",
      "5               0  0  1  0  0  0  0  0  0  0  0  2  0  0  0\n",
      "6               1  0  0  0  0  1  0  0  0  0  0  0  0  0  0\n",
      "7               0  1  0  0  0  1  0  0  1  0  0  0  0  0  1\n",
      "8               0  1  0  0  0  0  0  0  0  0  0  0  0  0  0\n",
      "9               1  0  0  0  2  0  0  0  2  2  0  0  0  1  0\n",
      "10              0  1  0  0  0  0  0  0  1  0  0  1  0  1  0\n",
      "11              0  0  1  0  0  1  0  0  0  0  0  1  0  0  0\n",
      "12              0  1  1  0  0  0  0  0  0  0  0  0  1  0  0\n",
      "13              0  1  0  2  0  0  2  3  0  0  1  0  0  0  0\n",
      "14              1  2  2  1  0  0  0  0  2  1  1  0  0  0  0\n",
      "15              0  0  0  0  0  0  1  0  0  3  0  0  0  0  0\n",
      "17              0  0  0  0  0  0  1  1  0  0  0  0  0  0  0\n",
      "18              0  0  0  0  0  0  0  1  0  0  0  0  0  0  0\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'GACGAACGCTGGCGGCGTGCCTAATACATGCAAGTCGAGCGGACCGACGGGAGCTTGCTCCCTTAGGTCAGCGGCGGACGGGTGAGTAACACGTGGGTAACCTGCCTGTAAGACTGGGATAACTCCGGGAAACCGGGGCTAATACCGGATGCTTGATTGAACCGCATGGTTCAATCATAAAAGGTGGCTTTTAGCTACCACTTACAGATGGACCCGCGGCGCATTAGCTAGTTGGTGAGGTAACGGCTCACCAAGGCGACGATGCGTAGCCGACCTGAGAGGGTGATCGGCCACACTGGGACTGAGACACGGCCCAGACTCCTACGGGAGGCAGCAGTAGGGAATCTTCCGCAATGGACGAAAGTCTGACGGAGCAACGCCGCGTGAGTGATGAAGGTTTTCGGATCGTAAAACTCTGTTGTTAGGGAAGAACAAGTACCGTTCGAATAGGGCGGTACCTTGACGGTACCTAACCAGAAAGCCACGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGCGCGCGCAGGCGGTTTCTTAAGTCTGATGTGAAAGCCCCCGGCTCAACCGGGGAGGGTCATTGGAAACTGGGGAACTTGAGTGCAGAAGAGGAGAGTGGAATTCCACGTGTAGCGGTGAAATGCGTAGAGATGTGGAGGAACACCAGTGGCGAAGGCGACTCTCTGGTCTGTAACTGACGCTGAGGCGCGAAAGCGTGGGGAGCGAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGAGTGCTAAGTGTTAGAGGGTTTCCGCCCTTTAGTGCTGCAGCAAACGCATTAAGCACTCCGCCTGGGGAGTACGGTCGCAAGACTGAAACTCAAAGGAATTGACGGGGGCCCGCACAAGCGGTGGAGCATGTGGTTTAATTCGAAGCAACGCGAAGAACCTTACCAGGTCTTGACATCCTCTGACAACCCTAGAGATAGGGCTTCCCCTTCGGGGGCAGAGTGACAGGTGGTGCATGGTTGTCGTCAGCTCGTGTCGTGAGATGTTGGGTTAAGTCCCGCAACGAGCGCAACCCTTGATCTTAGTTGCCAGCATTCAGTTGGGCACTCTAAGGTGACTGCCGGTGACAAACCGGAGGAAGGTGGGGATGACGTCAAATCATCATGCCCCTTATGACCTGGGCTACACACGTGCTACAATGGGCAGAACAAAGGGCAGCGAAGCCGCGAGGCTAAGCCAATCCCACAAATCTGTTCTCAGTTCGGATCGCAGTCTGCAACTCGACTGCGTGAAGCTGGAATCGCTAGTAATCGCGGATCAGCATGCCGCGGTGAATACGTTCCCGGGCCTTGTACACACCGCCCGTCACACCACGAGAGTTTGTAACACCCGAAGTCGGTGAGGTAACCTTTTGGAGCCAGCCGCCGAAGGTGGGACAGATGATTGGGGTGAAGTCGTAACAAGGTAACCGCACATAGTCTAGAGCAATACGTAACTGAACGAAGTAGCAGGACATAACA'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>forward_primer</th>\n",
       "      <th>reverse_primer</th>\n",
       "      <th>descriptor</th>\n",
       "      <th>consensus_sequence</th>\n",
       "      <th>16S_seq_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>E</td>\n",
       "      <td>9</td>\n",
       "      <td>consensus_E_9_0_0(393)</td>\n",
       "      <td>GACGAACGCTGGCGGCGTGCCTAATACATGCAAGTCGAGCGGACCG...</td>\n",
       "      <td>1541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>E</td>\n",
       "      <td>9</td>\n",
       "      <td>consensus_E_9_0_1(111)</td>\n",
       "      <td>ATTGAACGCTAGCGGGATGCTTTACACATGCAAGTCGAACGGCAGC...</td>\n",
       "      <td>1531</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   forward_primer  reverse_primer              descriptor  \\\n",
       "27              E               9  consensus_E_9_0_0(393)   \n",
       "28              E               9  consensus_E_9_0_1(111)   \n",
       "\n",
       "                                   consensus_sequence  16S_seq_length  \n",
       "27  GACGAACGCTGGCGGCGTGCCTAATACATGCAAGTCGAGCGGACCG...            1541  \n",
       "28  ATTGAACGCTAGCGGGATGCTTTACACATGCAAGTCGAACGGCAGC...            1531  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'GACGAACGCTGGCGGCGTGCCTAATACATGCAAGTCGAGCGGACCGACGGGAGCTTGCTCCCTTAGGTCAGCGGCGGACGGGTGAGTAACACGTGGGTAACCTGCCTGTAAGACTGGGATAACTCCGGGAAACCGGGGCTAATACCGGATGCTTGATTGAACCGCATGGTTCAATCATAAAAGGTGGCTTTTAGCTACCACTTACAGATGGACCCGCGGCGCATTAGCTAGTTGGTGAGGTAACGGCTCACCAAGGCGACGATGCGTAGCCGACCTGAGAGGGTGATCGGCCACACTGGGACTGAGACACGGCCCAGACTCCTACGGGAGGCAGCAGTAGGGAATCTTCCGCAATGGACGAAAGTCTGACGGAGCAACGCCGCGTGAGTGATGAAGGTTTTCGGATCGTAAAACTCTGTTGTTAGGGAAGAACAAGTACCGTTCGAATAGGGCGGTACCTTGACGGTACCTAACCAGAAAGCCACGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGCGCGCGCAGGCGGTTTCTTAAGTCTGATGTGAAAGCCCCCGGCTCAACCGGGGAGGGTCATTGGAAACTGGGGAACTTGAGTGCAGAAGAGGAGAGTGGAATTCCACGTGTAGCGGTGAAATGCGTAGAGATGTGGAGGAACACCAGTGGCGAAGGCGACTCTCTGGTCTGTAACTGACGCTGAGGCGCGAAAGCGTGGGGAGCGAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGAGTGCTAAGTGTTAGAGGGTTTCCGCCCTTTAGTGCTGCAGCAAACGCATTAAGCACTCCGCCTGGGGAGTACGGTCGCAAGACTGAAACTCAAAGGAATTGACGGGGGCCCGCACAAGCGGTGGAGCATGTGGTTTAATTCGAAGCAACGCGAAGAACCTTACCAGGTCTTGACATCCTCTGACAACCCTAGAGATAGGGCTTCCCCTTCGGGGGCAGAGTGACAGGTGGTGCATGGTTGTCGTCAGCTCGTGTCGTGAGATGTTGGGTTAAGTCCCGCAACGAGCGCAACCCTTGATCTTAGTTGCCAGCATTCAGTTGGGCACTCTAAGGTGACTGCCGGTGACAAACCGGAGGAAGGTGGGGATGACGTCAAATCATCATGCCCCTTATGACCTGGGCTACACACGTGCTACAATGGGCAGAACAAAGGGCAGCGAAGCCGCGAGGCTAAGCCAATCCCACAAATCTGTTCTCAGTTCGGATCGCAGTCTGCAACTCGACTGCGTGAAGCTGGAATCGCTAGTAATCGCGGATCAGCATGCCGCGGTGAATACGTTCCCGGGCCTTGTACACACCGCCCGTCACACCACGAGAGTTTGTAACACCCGAAGTCGGTGAGGTAACCTTTTGGAGCCAGCCGCCGAAGGTGGGACAGATGATTGGGGTGAAGTCGTAACAAGGTAACCGCACATAGTCTAGAGCAATACGTAACTGAACGAAGTAGCAGGACATAACA'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'ATTGAACGCTAGCGGGATGCTTTACACATGCAAGTCGAACGGCAGCGCGAAGAAAGCTTGCTTTCTTTGGCGGCGAGTGGCGAACGGGTGAGTAATATATCGGAACGTGCCCAGTAGCGGGGGATAACTACGCGAAAGCGTGGCTAATACCGCATACGCCCTACGGGGGAAAGGGGGGGATCGCAAGACCTCTCACTATTGGAGCGGCCGATACCAGATTAGCTTGTTGGTGAGGTAAAGGCTCACCAAGGCCGCGATCTGTAGCTGGTTTGAGAGGACGACCAGCCACACTGGGACTGAGACACGGCCCAGACTCCTACGGGAGGCAGCAGTGGGGAATTTTGGACAATGGGGGAAACCCTGATCCAGCCATCCCGCGTGTGCGATGAAGGCCTTCGGGTTGTAAAGCACTTTTGGCAGGGAAGAAAAGGCGCCGGATAATACCTGGCGCTGATGACGGTACCTGCAGAATAAGCACCGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGTGCGCAGGCGGTTCGGAAAGAAAGATGTGAAATCCCAGAGCTTAACTTTGGAACTGCATTTTTAACTGCCGAGCTAGAGTATGTCAGAGGGGGGTAGAATTCCACGTGTAGCAGTGAAATGCGTAGAGATGTGGAGGAATACCGATGGCGAAGGCAGCCCCCTGGGATAATACTGACGCTCATGCACGAAAGCGTGGGGAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCCTAAACGATGTCAACTAGCTGTTGGGGCCTTCGGGCCTTAGTAGCGCAGCTAACGCGTGAAGTTGACCGCCTGGGGAGTACGGTCGCAAGATTAAAACTCAAAGGAATTGACGGGGACCCGCACAAGCGGTGGATGATGTGGATTAATTCGATGCAACGCGAAAAACCTTACCTACCCTTGACATGTCTGGAAGCTTCGAGAGATTGAAGTGTGCTCGCAAGAGAACCGGAACACAGGTGCTGCATGGCTGTCGTCAGCTCGTGTCGTGAGATGTTGGGTTAAGTCCCGCAACGAGCGCAACCCTTGTCATTAGTTGCTACGAAAGGGCACTCTAATGAGACTGCCGGTGACAAACCGGAGGAAGGTGGGGATGACGTCAAGTCCTCATGGCCCTTATGGGTAGGGCTTCACACGTCATACAATGGTCGGGACAGAGGGTCGCCAACCCGCGAGGGGGAGCCAATCCCAGAAACCCGATCGTAGTCCGGATTGCAGGCTGCAACTCGCCTGCATGAAGTCGGAATCGCTAGTAATCGCGGATCAGACATGTCGCGGTGAATACGTTCCCGGGTCTTGTACACACCGCCCGTCACACCATGGGAGTGGGTTTTACCAGAAGTAGTTAGCCTAACCGTCAAGGAGGGCGATTACCACGGTAGGATTCATGACTGGGGTGAAGTCGTAACAAGGTAACCGCACATAGTCTAGAGCAATACGTAACTGAACGAAGTACAGGACATAACA'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>forward_primer</th>\n",
       "      <th>reverse_primer</th>\n",
       "      <th>descriptor</th>\n",
       "      <th>consensus_sequence</th>\n",
       "      <th>16S_seq_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>E</td>\n",
       "      <td>9</td>\n",
       "      <td>consensus_E_9_0_0(393)</td>\n",
       "      <td>GACGAACGCTGGCGGCGTGCCTAATACATGCAAGTCGAGCGGACCG...</td>\n",
       "      <td>1541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>E</td>\n",
       "      <td>9</td>\n",
       "      <td>consensus_E_9_0_1(111)</td>\n",
       "      <td>ATTGAACGCTAGCGGGATGCTTTACACATGCAAGTCGAACGGCAGC...</td>\n",
       "      <td>1531</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   forward_primer  reverse_primer              descriptor  \\\n",
       "27              E               9  consensus_E_9_0_0(393)   \n",
       "28              E               9  consensus_E_9_0_1(111)   \n",
       "\n",
       "                                   consensus_sequence  16S_seq_length  \n",
       "27  GACGAACGCTGGCGGCGTGCCTAATACATGCAAGTCGAGCGGACCG...            1541  \n",
       "28  ATTGAACGCTAGCGGGATGCTTTACACATGCAAGTCGAACGGCAGC...            1531  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df=pd.read_csv(\"class_16S_sequences.csv\")\n",
    "ct = pd.crosstab(df['reverse_primer'], df['forward_primer'])\n",
    "print(ct)\n",
    "consensus_sequence=df.query(\"forward_primer== 'E' and reverse_primer== 9\")\n",
    "seq_list= consensus_sequence['consensus_sequence'].tolist()\n",
    "display(seq_list[0])\n",
    "display(consensus_sequence)\n",
    "\n",
    "# Display both sequences if there are more than one\n",
    "if len(seq_list) > 0:\n",
    "    display(seq_list[0])  # First sequence\n",
    "if len(seq_list) > 1:\n",
    "    display(seq_list[1])  # Second sequence\n",
    "\n",
    "# Display the filtered consensus_sequence DataFrame\n",
    "display(consensus_sequence)\n",
    "\n",
    "#Your code here"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "231723e1",
   "metadata": {},
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73cdc167",
   "metadata": {},
   "source": [
    "### Task 2: Loading the reference 16S sequence database\n",
    "\n",
    "2.1> Load the 16S_reference.csv database file into a Pandas dataframe (use a different dataframe name to that you used when loading your sequences). This is a large version of the 'mini' dataframe you used last week. \n",
    "\n",
    "2.2> How many sequences are in the database?\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4924f7f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15155\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>GAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGTC...</td>\n",
       "      <td>Mycobacterium_heidelbergense(AJ000684)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bacteria;Proteobacteria;Gammaproteobacteria;Vi...</td>\n",
       "      <td>GTTTGATCCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...</td>\n",
       "      <td>Vibrio_atlanticus(EF599163)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>TAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAAC...</td>\n",
       "      <td>Mycobacterium_aubagnense(AY859683)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bacteria;Proteobacteria;Alphaproteobacteria;Rh...</td>\n",
       "      <td>AAGAGTTTGATCCTGGCTCAGAGCGAACGCTGGCGGCATGCTTAAC...</td>\n",
       "      <td>Acetobacter_senegalensis(AY883036)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bacteria;Proteobacteria;Betaproteobacteria;Bur...</td>\n",
       "      <td>ATTGAACGCTGGCGGCATGCCTTACACATGCAAGTCGAACGGTAAC...</td>\n",
       "      <td>Aquincola_tertiaricarbonis(DQ656489)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15150</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>AGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAACACA...</td>\n",
       "      <td>Leucobacter_komagatae(NR_112021.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15151</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>GATGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAG...</td>\n",
       "      <td>Thermobifida_fusca(NR_112015.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15152</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAA...</td>\n",
       "      <td>Microbispora_mesophila(NR_112014.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15153</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGAAAA...</td>\n",
       "      <td>Parvopolyspora_pallida(NR_112013.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15154</th>\n",
       "      <td>Bacteria;Proteobacteria;Gammaproteobacteria;En...</td>\n",
       "      <td>ACGCTGGCGGCAGGCCTAACACATGCAAGTCGAACGGTAGCACAGa...</td>\n",
       "      <td>Klebsiella_oxytoca(NR_112010.1)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15155 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      id  \\\n",
       "0      Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "1      Bacteria;Proteobacteria;Gammaproteobacteria;Vi...   \n",
       "2      Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "3      Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   \n",
       "4      Bacteria;Proteobacteria;Betaproteobacteria;Bur...   \n",
       "...                                                  ...   \n",
       "15150  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "15151  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "15152  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "15153  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "15154  Bacteria;Proteobacteria;Gammaproteobacteria;En...   \n",
       "\n",
       "                                                sequence  \\\n",
       "0      GAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGTC...   \n",
       "1      GTTTGATCCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...   \n",
       "2      TAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAAC...   \n",
       "3      AAGAGTTTGATCCTGGCTCAGAGCGAACGCTGGCGGCATGCTTAAC...   \n",
       "4      ATTGAACGCTGGCGGCATGCCTTACACATGCAAGTCGAACGGTAAC...   \n",
       "...                                                  ...   \n",
       "15150  AGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAACACA...   \n",
       "15151  GATGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAG...   \n",
       "15152  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAA...   \n",
       "15153  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGAAAA...   \n",
       "15154  ACGCTGGCGGCAGGCCTAACACATGCAAGTCGAACGGTAGCACAGa...   \n",
       "\n",
       "                                      species  \n",
       "0      Mycobacterium_heidelbergense(AJ000684)  \n",
       "1                 Vibrio_atlanticus(EF599163)  \n",
       "2          Mycobacterium_aubagnense(AY859683)  \n",
       "3          Acetobacter_senegalensis(AY883036)  \n",
       "4        Aquincola_tertiaricarbonis(DQ656489)  \n",
       "...                                       ...  \n",
       "15150      Leucobacter_komagatae(NR_112021.1)  \n",
       "15151         Thermobifida_fusca(NR_112015.1)  \n",
       "15152     Microbispora_mesophila(NR_112014.1)  \n",
       "15153     Parvopolyspora_pallida(NR_112013.1)  \n",
       "15154         Klebsiella_oxytoca(NR_112010.1)  \n",
       "\n",
       "[15155 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Your code here\n",
    "df1=pd.read_csv('16S_reference.csv')\n",
    "\n",
    "print(len(df1))\n",
    "display(df1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "710f0100",
   "metadata": {},
   "source": [
    "\n",
    "### Task 3: Searching for database matches\n",
    "\n",
    "3.1> For each query 16S sequence (i.e. for each of your own unknown 16S sequences), create a new column in the Pandas dataframe that represents the Levenshtein distance between that sequence and the sequences in the dataframe. The code you developed at the end of the week 7 workshop is close to what you will need to use here. Make sure that you create a different column name for the distance to each 16S sequence youre comparing to, for example 'dist_to_seq_1', 'dist_to_seq_2', etc. You can do this by run each of your group's sequences in turn, or each person can analyse a different one. If you're looking for an additional challenge, you could use a loop to process all sequences in your list of your group's sequences.\n",
    "\n",
    "3.2> Identify the best match(es) for each sequence: how close is the match? Hint: to find the closest match, you'll need to identify which row has the minimum value in the respective distance column. You then need to use .query() to return the row with that value.  You can adapt the code you developed in the Week 8 exercise notebook to do this. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "82b33951",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting levenshtein\n",
      "  Using cached levenshtein-0.27.1-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (160 kB)\n",
      "Collecting rapidfuzz<4.0.0,>=3.9.0\n",
      "  Using cached rapidfuzz-3.12.2-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)\n",
      "Installing collected packages: rapidfuzz, levenshtein\n",
      "  Attempting uninstall: rapidfuzz\n",
      "    Found existing installation: rapidfuzz 3.2.0\n",
      "    Uninstalling rapidfuzz-3.2.0:\n",
      "      Successfully uninstalled rapidfuzz-3.2.0\n",
      "Successfully installed levenshtein-0.27.1 rapidfuzz-3.12.2\n"
     ]
    }
   ],
   "source": [
    "!pip install levenshtein\n",
    "#pip is a package manager:this command installs the levenshtein package on our Noteable instance."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "66cb3f0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>GAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGTC...</td>\n",
       "      <td>Mycobacterium_heidelbergense(AJ000684)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bacteria;Proteobacteria;Gammaproteobacteria;Vi...</td>\n",
       "      <td>GTTTGATCCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...</td>\n",
       "      <td>Vibrio_atlanticus(EF599163)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>TAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAAC...</td>\n",
       "      <td>Mycobacterium_aubagnense(AY859683)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bacteria;Proteobacteria;Alphaproteobacteria;Rh...</td>\n",
       "      <td>AAGAGTTTGATCCTGGCTCAGAGCGAACGCTGGCGGCATGCTTAAC...</td>\n",
       "      <td>Acetobacter_senegalensis(AY883036)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bacteria;Proteobacteria;Betaproteobacteria;Bur...</td>\n",
       "      <td>ATTGAACGCTGGCGGCATGCCTTACACATGCAAGTCGAACGGTAAC...</td>\n",
       "      <td>Aquincola_tertiaricarbonis(DQ656489)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGAAAG...</td>\n",
       "      <td>Couchioplanes_caeruleus(D85479)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Bacteria;Proteobacteria;Alphaproteobacteria;Rh...</td>\n",
       "      <td>ATGGCTCAGAACGAACGCTGGCGGCAGGCCTAACACATGCAAGTCG...</td>\n",
       "      <td>Maritimibacter_alkaliphilus(DQ915443)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>TCTGTAAGTAGCAGGAAAGGCCCTTCGGGGTACTCGAGCGGCGAAC...</td>\n",
       "      <td>Nocardioides_fonticola(EF626689)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Bacteria;Proteobacteria;Gammaproteobacteria;Al...</td>\n",
       "      <td>AGGCCTAACACATGCAAGTCGAGCGGTAACACAAGGGAGCTTGCTC...</td>\n",
       "      <td>Shewanella_schlegeliana(AB081760)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAG...</td>\n",
       "      <td>Lentzea_albida(AB006176)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  id  \\\n",
       "0  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "1  Bacteria;Proteobacteria;Gammaproteobacteria;Vi...   \n",
       "2  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "3  Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   \n",
       "4  Bacteria;Proteobacteria;Betaproteobacteria;Bur...   \n",
       "5  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "6  Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   \n",
       "7  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "8  Bacteria;Proteobacteria;Gammaproteobacteria;Al...   \n",
       "9  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  GAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGTC...   \n",
       "1  GTTTGATCCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...   \n",
       "2  TAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAAC...   \n",
       "3  AAGAGTTTGATCCTGGCTCAGAGCGAACGCTGGCGGCATGCTTAAC...   \n",
       "4  ATTGAACGCTGGCGGCATGCCTTACACATGCAAGTCGAACGGTAAC...   \n",
       "5  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGAAAG...   \n",
       "6  ATGGCTCAGAACGAACGCTGGCGGCAGGCCTAACACATGCAAGTCG...   \n",
       "7  TCTGTAAGTAGCAGGAAAGGCCCTTCGGGGTACTCGAGCGGCGAAC...   \n",
       "8  AGGCCTAACACATGCAAGTCGAGCGGTAACACAAGGGAGCTTGCTC...   \n",
       "9  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAG...   \n",
       "\n",
       "                                  species  \n",
       "0  Mycobacterium_heidelbergense(AJ000684)  \n",
       "1             Vibrio_atlanticus(EF599163)  \n",
       "2      Mycobacterium_aubagnense(AY859683)  \n",
       "3      Acetobacter_senegalensis(AY883036)  \n",
       "4    Aquincola_tertiaricarbonis(DQ656489)  \n",
       "5         Couchioplanes_caeruleus(D85479)  \n",
       "6   Maritimibacter_alkaliphilus(DQ915443)  \n",
       "7        Nocardioides_fonticola(EF626689)  \n",
       "8       Shewanella_schlegeliana(AB081760)  \n",
       "9                Lentzea_albida(AB006176)  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1c6357a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>species</th>\n",
       "      <th>distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>GAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGTC...</td>\n",
       "      <td>Mycobacterium_heidelbergense(AJ000684)</td>\n",
       "      <td>387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bacteria;Proteobacteria;Gammaproteobacteria;Vi...</td>\n",
       "      <td>GTTTGATCCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...</td>\n",
       "      <td>Vibrio_atlanticus(EF599163)</td>\n",
       "      <td>403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>TAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAAC...</td>\n",
       "      <td>Mycobacterium_aubagnense(AY859683)</td>\n",
       "      <td>388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bacteria;Proteobacteria;Alphaproteobacteria;Rh...</td>\n",
       "      <td>AAGAGTTTGATCCTGGCTCAGAGCGAACGCTGGCGGCATGCTTAAC...</td>\n",
       "      <td>Acetobacter_senegalensis(AY883036)</td>\n",
       "      <td>473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bacteria;Proteobacteria;Betaproteobacteria;Bur...</td>\n",
       "      <td>ATTGAACGCTGGCGGCATGCCTTACACATGCAAGTCGAACGGTAAC...</td>\n",
       "      <td>Aquincola_tertiaricarbonis(DQ656489)</td>\n",
       "      <td>412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGAAAG...</td>\n",
       "      <td>Couchioplanes_caeruleus(D85479)</td>\n",
       "      <td>364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Bacteria;Proteobacteria;Alphaproteobacteria;Rh...</td>\n",
       "      <td>ATGGCTCAGAACGAACGCTGGCGGCAGGCCTAACACATGCAAGTCG...</td>\n",
       "      <td>Maritimibacter_alkaliphilus(DQ915443)</td>\n",
       "      <td>434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>TCTGTAAGTAGCAGGAAAGGCCCTTCGGGGTACTCGAGCGGCGAAC...</td>\n",
       "      <td>Nocardioides_fonticola(EF626689)</td>\n",
       "      <td>415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Bacteria;Proteobacteria;Gammaproteobacteria;Al...</td>\n",
       "      <td>AGGCCTAACACATGCAAGTCGAGCGGTAACACAAGGGAGCTTGCTC...</td>\n",
       "      <td>Shewanella_schlegeliana(AB081760)</td>\n",
       "      <td>417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAG...</td>\n",
       "      <td>Lentzea_albida(AB006176)</td>\n",
       "      <td>345</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  id  \\\n",
       "0  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "1  Bacteria;Proteobacteria;Gammaproteobacteria;Vi...   \n",
       "2  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "3  Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   \n",
       "4  Bacteria;Proteobacteria;Betaproteobacteria;Bur...   \n",
       "5  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "6  Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   \n",
       "7  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "8  Bacteria;Proteobacteria;Gammaproteobacteria;Al...   \n",
       "9  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  GAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGTC...   \n",
       "1  GTTTGATCCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...   \n",
       "2  TAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAAC...   \n",
       "3  AAGAGTTTGATCCTGGCTCAGAGCGAACGCTGGCGGCATGCTTAAC...   \n",
       "4  ATTGAACGCTGGCGGCATGCCTTACACATGCAAGTCGAACGGTAAC...   \n",
       "5  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGAAAG...   \n",
       "6  ATGGCTCAGAACGAACGCTGGCGGCAGGCCTAACACATGCAAGTCG...   \n",
       "7  TCTGTAAGTAGCAGGAAAGGCCCTTCGGGGTACTCGAGCGGCGAAC...   \n",
       "8  AGGCCTAACACATGCAAGTCGAGCGGTAACACAAGGGAGCTTGCTC...   \n",
       "9  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAG...   \n",
       "\n",
       "                                  species  distance  \n",
       "0  Mycobacterium_heidelbergense(AJ000684)       387  \n",
       "1             Vibrio_atlanticus(EF599163)       403  \n",
       "2      Mycobacterium_aubagnense(AY859683)       388  \n",
       "3      Acetobacter_senegalensis(AY883036)       473  \n",
       "4    Aquincola_tertiaricarbonis(DQ656489)       412  \n",
       "5         Couchioplanes_caeruleus(D85479)       364  \n",
       "6   Maritimibacter_alkaliphilus(DQ915443)       434  \n",
       "7        Nocardioides_fonticola(EF626689)       415  \n",
       "8       Shewanella_schlegeliana(AB081760)       417  \n",
       "9                Lentzea_albida(AB006176)       345  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from Levenshtein import distance\n",
    "string_to_compare = seq_list[0]\n",
    "df1['distance'] = df1['sequence'].apply(distance, args=(string_to_compare,))\n",
    "\n",
    "display(df1.head(10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "100c1f21",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['id', 'sequence', 'species', 'distance'], dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>species</th>\n",
       "      <th>distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9440</th>\n",
       "      <td>Bacteria;Firmicutes;Bacilli;Bacillales;Bacilla...</td>\n",
       "      <td>GACGAACGCTGGCGGCGTGCCTAATACATGCAAGTCGAGCGGACAG...</td>\n",
       "      <td>Bacillus_atrophaeus(AB021181)</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     id  \\\n",
       "9440  Bacteria;Firmicutes;Bacilli;Bacillales;Bacilla...   \n",
       "\n",
       "                                               sequence  \\\n",
       "9440  GACGAACGCTGGCGGCGTGCCTAATACATGCAAGTCGAGCGGACAG...   \n",
       "\n",
       "                            species  distance  \n",
       "9440  Bacillus_atrophaeus(AB021181)        80  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print(df1.columns)\n",
    "# Find the minimum distance value\n",
    "min_distance = df1['distance'].min()\n",
    "\n",
    "# Return the row with that minimum distance\n",
    "best_match = df1.query(\"distance == @min_distance\")\n",
    "\n",
    "display(best_match)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4b503d3",
   "metadata": {},
   "source": [
    "# Task 4: Visualising your analysis results\n",
    "\n",
    "4.1> Run a needleman_wunsch alignment between your 16S sequence(s) and the best match. Python code to run an Needleman Wunsch alignment is shown below. You'll need to extract your own 16S sequence and the best match 16S sequence as strings before you can align them. \n",
    "4.2> View the alignment. How well is the sequence aligned? Can you see poorer alignment at the ends of the sequence caused by the presence of PCR primers in your sequence?\n",
    "\n",
    "Hint: If you're struggling with getting the sequence to convert to a string and you are seeing the text 'Name: sequence, dtype: object' when you print it, try suffixing the expression with .iloc[0]. This extracts the string from the Pandas data series.\n",
    "\n",
    "An example code for running a Needleman-Wunsch alignment is shown below:\n",
    "\n",
    "```python\n",
    "import needleman_wunsch\n",
    "seq1='ATGCTGAGCTAGCGGCTATATTCTATCGGGAGCGATTTACTACTC'\n",
    "seq2='ATGCTAGGTAGCGGACTATATACTATCGCGAGCGATTAACTAGCC'\n",
    "print(needleman_wunsch.align(seq1, seq2))\n",
    "```\n",
    "Expected output:\n",
    "```\n",
    "seq1: ATGCTGAGCTAGCGG-CTATATTCTATCGGGAGCGATTTACTA-CTC\n",
    "      ||||| || |||||| |||||| |||||| |||||||| |||| | |\n",
    "seq2: ATGCT-AGGTAGCGGACTATATACTATCGCGAGCGATTAACTAGC-C\n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "77cd6224",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "seq1: GACGAACGCTGGCGGCGTGCCTAATACATGCAAGTCGAGCGGACCGACGGGAGCTTGCTCCCTTAGGTCA\n",
      "      |||||||||||||||||||||||||||||||||||||||||||| || ||||||||||||||| | || |\n",
      "seq2: GACGAACGCTGGCGGCGTGCCTAATACATGCAAGTCGAGCGGACAGATGGGAGCTTGCTCCCTGATGTTA\n",
      "\n",
      "seq1: GCGGCGGACGGGTGAGTAACACGTGGGTAACCTGCCTGTAAGACTGGGATAACTCCGGGAAACCGGGGCT\n",
      "      ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "seq2: GCGGCGGACGGGTGAGTAACACGTGGGTAACCTGCCTGTAAGACTGGGATAACTCCGGGAAACCGGGGCT\n",
      "\n",
      "seq1: AATACCGGATGCTTGATTGAACCGCATGGTTCAATCATAAAAGGTGGCTTTTAGCTACCACTTACAGATG\n",
      "      ||||||||||||||| |||||||||||||||||| ||||||||||||| ||  |||||||||||||||||\n",
      "seq2: AATACCGGATGCTTGTTTGAACCGCATGGTTCAAACATAAAAGGTGGC-TTCGGCTACCACTTACAGATG\n",
      "\n",
      "seq1: GACCCGCGGCGCATTAGCTAGTTGGTGAGGTAACGGCTCACCAAGGCGACGATGCGTAGCCGACCTGAGA\n",
      "      ||||||||||||||||||||||||||||||||||||||||||||||| ||||||||||||||||||||||\n",
      "seq2: GACCCGCGGCGCATTAGCTAGTTGGTGAGGTAACGGCTCACCAAGGCAACGATGCGTAGCCGACCTGAGA\n",
      "\n",
      "seq1: GGGTGATCGGCCACACTGGGACTGAGACACGGCCCAGACTCCTACGGGAGGCAGCAGTAGGGAATCTTCC\n",
      "      ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "seq2: GGGTGATCGGCCACACTGGGACTGAGACACGGCCCAGACTCCTACGGGAGGCAGCAGTAGGGAATCTTCC\n",
      "\n",
      "seq1: GCAATGGACGAAAGTCTGACGGAGCAACGCCGCGTGAGTGATGAAGGTTTTCGGATCGTAAAACTCTGTT\n",
      "      |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||| |||||||\n",
      "seq2: GCAATGGACGAAAGTCTGACGGAGCAACGCCGCGTGAGTGATGAAGGTTTTCGGATCGTAAAGCTCTGTT\n",
      "\n",
      "seq1: GTTAGGGAAGAACAAGTACCGTTCGAATAGGGCGGTACCTTGACGGTACCTAACCAGAAAGCCACGGCTA\n",
      "      ||||||||||||||||| |||||| |||||||||| ||||||||||||||||||||||||||||||||||\n",
      "seq2: GTTAGGGAAGAACAAGTGCCGTTCAAATAGGGCGGCACCTTGACGGTACCTAACCAGAAAGCCACGGCTA\n",
      "\n",
      "seq1: ACTACGTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGCGCG\n",
      "      |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||| || \n",
      "seq2: ACTACGTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGGGCT\n",
      "\n",
      "seq1: CGCAGGCGGTTTCTTAAGTCTGATGTGAAAGCCCCCGGCTCAACCGGGGAGGGTCATTGGAAACTGGGGA\n",
      "      |||||||||||||||||||||||||||||||||   ||||||||||||||||||||||||||||||||||\n",
      "seq2: CGCAGGCGGTTTCTTAAGTCTGATGTGAAAGCCcccGGCTCAACCGGGGAGGGTCATTGGAAACTGGGGA\n",
      "\n",
      "seq1: ACTTGAGTGCAGAAGAGGAGAGTGGAATTCCACGTGTAGCGGTGAAATGCGTAGAGATGTGGAGGAACAC\n",
      "      ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "seq2: ACTTGAGTGCAGAAGAGGAGAGTGGAATTCCACGTGTAGCGGTGAAATGCGTAGAGATGTGGAGGAACAC\n",
      "\n",
      "seq1: CAGTGGCGAAGGCGACTCTCTGGTCTGTAACTGACGCTGAGGCGCGAAAGCGTGGGGAGCGAACAGGATT\n",
      "      |||||||||||||||||    ||||||||||||||||||||| |||||||||||||||||||||||||||\n",
      "seq2: CAGTGGCGAAGGCGACTctctGGTCTGTAACTGACGCTGAGGAGCGAAAGCGTGGGGAGCGAACAGGATT\n",
      "\n",
      "seq1: AGATACCCTGGTAGTCCACGCCGTAAACGATGAGTGCTAAGTGTTAGAG-GGTTTCCGCCCTTTAGTGCT\n",
      "      ||||||||||||||||||||||||||||||||||||||||||||||| |   ||||||||| ||||||||\n",
      "seq2: AGATACCCTGGTAGTCCACGCCGTAAACGATGAGTGCTAAGTGTTAG-GgggTTTCCGCCCCTTAGTGCT\n",
      "\n",
      "seq1: GCAGCAAACGCATTAAGCACTCCGCCTGGGGAGTACGGTCGCAAGACTGAAACTCAAAGGAATTGACGGG\n",
      "      ||||| ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||| \n",
      "seq2: GCAGCTAACGCATTAAGCACTCCGCCTGGGGAGTACGGTCGCAAGACTGAAACTCAAAGGAATTGACGGg\n",
      "\n",
      "seq1: GGCCCGCACAAGCGGTGGAGCATGTGGTTTAATTCGAAGCAACGCGAAGAACCTTACCAGGTCTTGACAT\n",
      "        ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "seq2: ggCCCGCACAAGCGGTGGAGCATGTGGTTTAATTCGAAGCAACGCGAAGAACCTTACCAGGTCTTGACAT\n",
      "\n",
      "seq1: CCTCTGACAACCCTAGAGATAGGGCTTCCCCTTCGGGGGCAGAGTGACAGGTGGTGCATGGTTGTCGTCA\n",
      "      ||||||||| ||||||||||||||||||||||||||   |||||||||||||||||||||||||||||||\n",
      "seq2: CCTCTGACACCCCTAGAGATAGGGCTTCCCCTTCGGgggCAGAGTGACAGGTGGTGCATGGTTGTCGTCA\n",
      "\n",
      "seq1: GCTCGTGTCGTGAGATGTTGGGTTAAGTCCCGCAACGAGCGCAACCCTTGATCTTAGTTGCCAGCATTCA\n",
      "      ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "seq2: GCTCGTGTCGTGAGATGTTGGGTTAAGTCCCGCAACGAGCGCAACCCTTGATCTTAGTTGCCAGCATTCA\n",
      "\n",
      "seq1: GTTGGGCACTCTAAGGTGACTGCCGGTGACAAACCGGAGGAAGGTGGGGATGACGTCAAATCATCATGCC\n",
      "      ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "seq2: GTTGGGCACTCTAAGGTGACTGCCGGTGACAAACCGGAGGAAGGTGGGGATGACGTCAAATCATCATGCC\n",
      "\n",
      "seq1: CCTTATGACCTGGGCTACACACGTGCTACAATGGGCAGAACAAAGGGCAGCGA-AGCCGCGAGGCTAAGC\n",
      "      ||||||||||||||||||    |||||||||||| |||||||||||||||||| | |||||||| |||||\n",
      "seq2: CCTTATGACCTGGGCTACacacGTGCTACAATGGACAGAACAAAGGGCAGCGAGA-CCGCGAGGTTAAGC\n",
      "\n",
      "seq1: CAATCCCACAAATCTGTTCTCAGTTCGGATCGCAGTCTGCAACTCGACTGCGTGAAGCTGGAATCGCTAG\n",
      "      ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "seq2: CAATCCCACAAATCTGTTCTCAGTTCGGATCGCAGTCTGCAACTCGACTGCGTGAAGCTGGAATCGCTAG\n",
      "\n",
      "seq1: TAATCGCGGATCAGCATGCCGCGGTGAATACGTTCCCGGGCCTTGTACACACCGCCCGTCACACCACGAG\n",
      "      ||||||||||||||||||||||||||||||||||||||||||||||||    ||||||||||||||||||\n",
      "seq2: TAATCGCGGATCAGCATGCCGCGGTGAATACGTTCCCGGGCCTTGTACacacCGCCCGTCACACCACGAG\n",
      "\n",
      "seq1: AGTTTGTAACACCCGAAGTCGGTGAGGTAACCTTT-TGGAGCCAGCCGCCGAAGGTGGGACAGATGATTG\n",
      "      ||||||||||||||||||||||||||||||||||| ||||||||||||||||||||||||||||||||||\n",
      "seq2: AGTTTGTAACACCCGAAGTCGGTGAGGTAACCTTTATGGAGCCAGCCGCCGAAGGTGGGACAGATGATTG\n",
      "\n",
      "seq1: GGGTGAAGTCGTAACAAGGTAACCGCACATAGTCTAGAGCAATACGTAACTGAACGAAGTAGCAGGACAT\n",
      "      |||||||||||||||||||| |  || | || ||  | |  | | |    ||  |   |  || ||  ||\n",
      "seq2: GGGTGAAGTCGTAACAAGGT-A--GC-CGTA-TC--G-G--A-A-G---GTG--C---G--GCTGG--AT\n",
      "\n",
      "seq1: AACA\n",
      "        ||\n",
      "seq2: --CA\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import needleman_wunsch\n",
    "best_match_sequence = best_match['sequence'].iloc[0]\n",
    "\n",
    "# Display the alignment results for both\n",
    "print(needleman_wunsch.align(seq_list[0], best_match_sequence))\n",
    "\n",
    "#Your code here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "676979a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "seq1: -ATTGAACGCTAGCGG-GATGCTTTACACATGCAAGTCGAACGGCAGCGCGAAGAAAGCTTGCTTTCTTT\n",
      "       |  ||||||| |||| | ||| | | ||||||||||||| ||| | |  || |  |||||||  ||  |\n",
      "seq2: GA-CGAACGCTGGCGGCG-TGCCTAATACATGCAAGTCGAGCGG-A-C-AGATGGGAGCTTGC--TCCCT\n",
      "\n",
      "seq1: GGCGGCGAGTGGCGAACGGGTGAGTAATATATCGG-AACGTGCCCAGT-AG-CGGGGGATAACTACGCGA\n",
      "       |  |  || |||| |||||||||||| |  | || ||| || || || || |  ||||||||| || ||\n",
      "seq2: -GATGTTAGCGGCGGACGGGTGAGTAACACGTGGGTAACCTG-CCTGTAAGAC-TGGGATAACTCCGGGA\n",
      "\n",
      "seq1: AAGCGTGGCTAATACCGCATAC------G--CC-CTACGG------GGGAAAGGGGGGGATC-GCAAGAC\n",
      "      || || ||||||||||| || |      |  || | | ||         ||| || ||  || ||   ||\n",
      "seq2: AACCGGGGCTAATACCGGATGCTTGTTTGAACCGC-ATGGTTCAAACATAAAAGGTGGCTTCGGC--TAC\n",
      "\n",
      "seq1: CTCTCACTA-TTGGA---GCGGCCGATACCAGATTAGCTTGTTGGTGAGGTAAAGGCTCACCAAGGCCGC\n",
      "      | || || |  ||||   |||| ||    |  ||||||| ||||||||||||| |||||||||||||  |\n",
      "seq2: CACTTAC-AGATGGACCCGCGG-CG----C--ATTAGCTAGTTGGTGAGGTAACGGCTCACCAAGGCAAC\n",
      "\n",
      "seq1: GAT-CTGTAGCTGGTTTGAGAGGACGACCAGCCACACTGGGACTGAGACACGGCCCAGACTCCTACGGGA\n",
      "      ||| | ||||| |   |||||||  || | ||||||||||||||||||||||||||||||||||||||||\n",
      "seq2: GATGC-GTAGCCGACCTGAGAGGGTGATCGGCCACACTGGGACTGAGACACGGCCCAGACTCCTACGGGA\n",
      "\n",
      "seq1: GGCAGCAGTGGGGAAT-TTTGGACAATGGGGGAAACCCTGATC-CAGCCATC-CCGCGTGTGCGATGAAG\n",
      "      ||||||||| |||||| ||  | ||||||  ||||  |||| |  || || | ||||||| | |||||||\n",
      "seq2: GGCAGCAGTAGGGAATCTTCCG-CAATGGACGAAAGTCTGA-CGGAG-CAACGCCGCGTGAGTGATGAAG\n",
      "\n",
      "seq1: GCCTTCGGGTTGTAAAGCACTTTTGGCAGGGAAGAA-AAGGCGCCG--GATAATA--CCTGGC-GCTGAT\n",
      "      |  ||||| | ||||||| || |||  ||||||||| || | ||||   | ||||   | |||  ||  |\n",
      "seq2: GTTTTCGGATCGTAAAGCTCTGTTGTTAGGGAAGAACAA-GTGCCGTTCA-AATAGGGC-GGCACCT--T\n",
      "\n",
      "seq1: GACGGTACCT--GCAGAATAAG-CACCGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGG-GTGCA\n",
      "      ||||||||||   ||| | ||| || |||||||||||||||||||||||||||||||||||||| | |||\n",
      "seq2: GACGGTACCTAACCAG-A-AAGCCA-CGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGTG-GCA\n",
      "\n",
      "seq1: AGCGTTAAT-CGGAATTACTGGGCGTAAAGCG-TGCGCAGGCGG-TTCGGAAAGAAAGATGTGAAATCC-\n",
      "      ||||||  | |||||||| ||||||||||| | | ||||||||| |||   |||   ||||||||| || \n",
      "seq2: AGCGTT-GTCCGGAATTATTGGGCGTAAAGGGCT-CGCAGGCGGTTTC-TTAAGTCTGATGTGAAAGCCc\n",
      "\n",
      "seq1: CAGAGCTTAACTTTGGA-ACTGCATTTTTAACTGCCGAGCTAGAGTATGTCAG-AG-GG-GGGTAGAATT\n",
      "        | ||| |||   |||   | ||||   |||||  || || |||  || ||| || || | || |||||\n",
      "seq2: ccG-GCTCAACCGGGGAGGGT-CATTGGAAACTGGGGAACTTGAG--TG-CAGAAGAGGAGAGTGGAATT\n",
      "\n",
      "seq1: CCACGTGTAGCAGTGAAATGCGTAGAGATGTGGAGGAATACC-GATGGCGAAGGC-AGCCCCCTGGGAT-\n",
      "      ||||||||||| |||||||||||||||||||||||||| ||| | |||||||||| | |     ||  | \n",
      "seq2: CCACGTGTAGCGGTGAAATGCGTAGAGATGTGGAGGAACACCAG-TGGCGAAGGCGA-CTctctGGTCTG\n",
      "\n",
      "seq1: AATACTGACGCTCATGCA-CGAAAGCGTGGGGAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCCTA\n",
      "       | ||||||||| | | | |||||||||||||||| ||||||||||||||||||||||||||||||| ||\n",
      "seq2: TA-ACTGACGCTGA-GGAGCGAAAGCGTGGGGAGCGAACAGGATTAGATACCCTGGTAGTCCACGCCGTA\n",
      "\n",
      "seq1: AACGAT--GT-CAACTAGCTGTT-GGGGCCTT-CGGGCCTTAGTAGC-GCAGCTAACGCGTGAAGTTGAC\n",
      "      ||||||  || | |  || |||| ||    || ||  ||||||| || ||||||||||| | |||    |\n",
      "seq2: AACGATGAGTGCTA--AG-TGTTAGGgggTTTCCGCCCCTTAGT-GCTGCAGCTAACGCATTAAGCACTC\n",
      "\n",
      "seq1: CGCCTGGGGAGTACGGTCGCAAGATTAAAACTCAAAGGAATTGACGGGGACCCGCACAAGCGGTGGATG-\n",
      "      |||||||||||||||||||||||| | ||||||||||||||||||||   ||||||||||||||||| | \n",
      "seq2: CGCCTGGGGAGTACGGTCGCAAGACTGAAACTCAAAGGAATTGACGGgggCCCGCACAAGCGGTGGA-GC\n",
      "\n",
      "seq1: ATGTGGATTAATTCGATGCAACGCGAAAAACCTTACCTA-CCCTTGACAT-GTCTGGA-AGCTTCGAGAG\n",
      "      |||||| ||||||||| |||||||||| ||||||||| |   ||||||||  ||| || | |  | ||||\n",
      "seq2: ATGTGGTTTAATTCGAAGCAACGCGAAGAACCTTACC-AGGTCTTGACATCCTCT-GACA-CCCCTAGAG\n",
      "\n",
      "seq1: ATTGAAGTGTGC-T-CGC-AAGAGAACCGGA-ACACAGGTGCTGCATGGCTGTCGTCAGCTCGTGTCGTG\n",
      "      | |  || | || | | |   | |   | ||   ||||||| ||||||| ||||||||||||||||||||\n",
      "seq2: A-T--AG-G-GCTTCCCCTTCG-GgggCAGAGTGACAGGTGGTGCATGGTTGTCGTCAGCTCGTGTCGTG\n",
      "\n",
      "seq1: AGATGTTGGGTTAAGTCCCGCAACGAGCGCAACCCTTG-TCATTAGTTGCTA-CGA---A--AGGGCACT\n",
      "      |||||||||||||||||||||||||||||||||||||| || |||||||| | | |   |   |||||||\n",
      "seq2: AGATGTTGGGTTAAGTCCCGCAACGAGCGCAACCCTTGATC-TTAGTTGCCAGC-ATTCAGTTGGGCACT\n",
      "\n",
      "seq1: CTAATGAGACTGCCGGTGACAAACCGGAGGAAGGTGGGGATGACGTCAAGTCCTCATGGCCCTTATG-GG\n",
      "      |||| | |||||||||||||||||||||||||||||||||||||||||| || ||||| ||||||||   \n",
      "seq2: CTAAGGTGACTGCCGGTGACAAACCGGAGGAAGGTGGGGATGACGTCAAATCATCATGCCCCTTATGACC\n",
      "\n",
      "seq1: TAGGGCTTCACACGT-CATACAATGGTCGGGACAGAGGGTC-GCCA-ACCCGCGAGGGGGAGCCAATCCC\n",
      "      | ||||| |    || | |||||||| | | ||| |||| | || | | ||||||||   ||||||||||\n",
      "seq2: T-GGGCTACacacGTGC-TACAATGGACAGAACAAAGGG-CAGCGAGA-CCGCGAGGTTAAGCCAATCCC\n",
      "\n",
      "seq1: AGAAACCCGATCGT-AGTCCGGATTGCAGGCTGCAACTCGCCTGCATGAAG-TCGGAATCGCTAGTAATC\n",
      "      | ||| | | || | ||| ||||| |||| |||||||||| |||| ||||| | ||||||||||||||||\n",
      "seq2: ACAAATCTGTTC-TCAGTTCGGATCGCAGTCTGCAACTCGACTGCGTGAAGCT-GGAATCGCTAGTAATC\n",
      "\n",
      "seq1: GCGGATCAGACATGTCGCGGTGAATACGTTCCCGGGTCTTGTACACACCGCCCGTCACACCATGGGAGTG\n",
      "      ||||||||| |||| ||||||||||||||||||||| |||||||    |||||||||||||| | |||| \n",
      "seq2: GCGGATCAG-CATGCCGCGGTGAATACGTTCCCGGGCCTTGTACacacCGCCCGTCACACCACGAGAGTT\n",
      "\n",
      "seq1: GGTTTTACCAGAAGT-AGTTAGCCTAACCGTCAAGGAGGGC-GATTACC-ACGGTAGGATTCATGACTGG\n",
      "       ||   ||| |||||  || ||  ||||| | | ||| | | |    || | ||| |||   |||| |||\n",
      "seq2: TGTAACACCCGAAGTCGGTGAG-GTAACCTTTATGGA-GCCAG-CCGCCGAAGGTGGGACAGATGATTGG\n",
      "\n",
      "seq1: GGTGAAGTCGTAACAAGGTAACCGCACATAGTCTAGAGCAATACGTAACTGAACGAAGTACAGGACATAA\n",
      "      ||||||||||||||||||| |  || | || ||  | |  | | |    ||  ||  |  | ||  ||  \n",
      "seq2: GGTGAAGTCGTAACAAGGT-A--GC-CGTA-TC--G-G--A-A-G---GTG--CG--G--CTGG--AT--\n",
      "\n",
      "seq1: CA\n",
      "      ||\n",
      "seq2: CA\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(needleman_wunsch.align(seq_list[1], best_match_sequence))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d53ba27",
   "metadata": {},
   "source": [
    "# Finding antibiotic producers from soil sample\n",
    "\n",
    "Using the same method shown in Task 1 to 4, the species of antiobiotic-producing bacteria found from my soil sample were identified. The sample was collected from Almond River, Edinburgh at coordinates 55°57'07.6\"N, 3°24'12.7\"W. The species' 16S rDNAs sequences were retrieved from PCR test conducted at Edinburgh University's laboratory. These sequences were then compared against the sequences in Ribosome Database Project. \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a1fc15a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "forward_primer  A  B  C  D  E  F  G  H  I  J  L  R  S  U  V  W  Z\n",
      "reverse_primer                                                   \n",
      "0               0  1  2  0  0  0  0  0  1  0  0  0  0  0  0  0  0\n",
      "1               0  0  1  1  0  1  0  1  0  0  0  0  0  0  0  0  0\n",
      "2               0  3  0  1  0  0  0  0  0  2  1  0  0  0  0  0  0\n",
      "3               0  0  2  1  1  1  1  0  0  3  0  0  0  0  0  0  1\n",
      "4               1  0  0  3  0  0  1  0  1  1  0  0  0  0  0  0  0\n",
      "5               0  0  1  0  0  0  0  0  2  1  0  0  0  0  0  0  0\n",
      "6               0  1  0  0  1  0  0  1  0  0  0  0  0  1  0  0  0\n",
      "7               1  0  0  0  0  0  0  0  0  1  0  0  0  0  0  0  0\n",
      "8               1  2  0  0  1  0  0  0  0  0  0  0  0  0  0  0  0\n",
      "9               0  2  0  0  0  0  1  0  0  1  0  0  0  0  0  0  0\n",
      "10              0  0  0  0  0  0  1  0  0  2  0  1  1  0  1  0  0\n",
      "11              0  1  0  0  1  0  1  0  1  0  0  0  0  0  0  0  0\n",
      "12              0  0  0  0  0  0  0  0  0  2  0  0  0  0  0  0  0\n",
      "16              1  0  0  1  1  1  1  1  1  1  0  0  0  0  1  1  0\n",
      "18              1  0  0  0  0  0  0  0  0  2  0  0  0  0  0  0  0\n",
      "19              0  0  0  0  0  0  0  0  0  2  0  0  0  0  0  0  0\n",
      "32              0  0  0  0  0  0  0  1  0  0  0  0  0  0  0  0  0\n",
      "33              0  0  1  0  0  0  0  0  0  0  0  0  0  0  0  0  0\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'TAAACCTCGTCGCGTGCTTAACACATGCAAGTCGAACGGTGAAGCCCAGCTTGCTGGGTGGATCAGTGGCGAACGGGTGAGTAACACGTGAGCAACCTGCCCCTGACTCTGGGATAAGCGCTGGAAACGGCGTCTAATACTGGATATGTGACGTGACCGCATGGTCTGCGTCTGAAAGAATTTCGGTTGGGGATGGGCTCGCGGCCTATCAGCTTGTTGGTGAGGTAATGGCTCACCAAGGCGTCGACGGGTAGCCGGCCTGAGAGGGTGACCGGCCACACTGGGACTGAGACACGGCCCAGACTCCTACGGGAGGCAGCAGTGGGGAATATTGCACAATGGGCGCAAGCCTGATGCAGCAACGCCGCGTGAGGGATGACGGCCTTCGGGTTGTAAACCTCTTTTAGCAGGGAAGAAGCGAAAGTGACGGTACCTGCAGAAAAAGCGCCGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGGCGCAAGCGTTATCCGGAATTATTCGGCATAAAGTCGTAGGCGGTTTGTCGCGTCTGCTGTGAAATCCGGAGGCTCAACCTCCGGCCTGCAGTGGGTACGGGCAGACTAGAGTGCGGTAGGGGAGTCAAATCCTGGTGTAGCGGTGGAATGCGCAGATATCAGGAGGAACACCGATGGCGAAGGCAGATCTCTGGGCCGTAACTGACGCTGAGGAGCGAAAGGGTGGGGAGCGGACGACTATACCCCTGGTAGTCCACCCCGTAAACGTTGGGAACTAGTTGTGGGGTCCATTCCACGGATTCCGTGACGCAGCTAACGCATTAAGACTCCCGCCTGGGGAGTACGGCCGCAAGGCTAAAACTCAAAGGAATTGACGGGGACCCGCACAAGCGGCGGAGCATGCGGATTAATTCGATGCAACGCGAAGAACCTTACCAAGGCTTGACATATACCGGAAAGCCGTAGAGATACGGCCCCCCTTGTGGTCGGTATACAGGTGGTGCATGGCTGTCGTCAGCTCGTGTCGTGAGATGTTGGGTTAAGTCCCGCAACGAGCGCAACCCCTATCTTATGTTGCCAGCGCGTTATGGCGGGGACTCGTTGCCAGGGGTCAACTCGGAGGAAGGTGGGGACGACGTCAAGTCATCATGCCCCTTATGTCCAGGGCTTCACACATGCTACAATGGCCAGTACAAGAAGGGCTGCGAGACCGTGAGGTGGAGCGAATCCCTTAAAGCTGGTCTCAGTTCGGATTCGGGGTCTGCAACTCGACCCATGAATGAGAAGTCACGCTTAGTAATCGCAGATCAGCAACGCTGCAGTGAATACGTTCCCGGGCCTTGTACACACCGCCCGTCAAGTCACGAAAGTTGGTAACACCCGAAGCCGGTGGCCTAACCCCTTGTGGGAGGGGAGCTGTCGAAGGTGGGACTGGCGATTGGGACTAAGTCGTAAGCGTAACCGTCGAGCTGCTAAGCAATGG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGATGAACCACTTCGGTGGGGATTAGTGGCGAACGGGTGAGTAACACGTGGGCAATCTGCCCTGCACTCTGGGACAAGCCCTGGAAACGGGGTCTAATACCGGATACTGACCCGCTTGGGCATCCAAGCGGTTCGAAAGCTCCGGCGGTGCAGGATGAGCCCGCGGCCTATCAGCTTGTTGGTGAGGTAATGGCTCACCAAGGCGACGACGGGTAGCCGGCCTGAGAGGGCGACCGGCCACACTGGGACTGAGACACGGCCCAGACTCCTACGGGAGGCAGCAGTGGGGAATATTGCACAATGGGCGAAAGCCTGATGCAGCGACGCCGCGTGAGGGATGACGGCCTTCGGGTTGTAAACCTCTTTCAGCAGGGAAGAAGCGAAAGTGACGGTACCTGCAGAAGAAGCGCCGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGGCGCGAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCTCGTAGGCGGCTTGTCACGTCGGTTGTGAAAGCCCGGGGCTTAACCCCGGGTCTGCAGTCGATACGGGCAGGCTAGAGTTCGGTAGGGGAGATCGGAATTCCTGGTGTAGCGGTGAAATGCGCAGATATCAGGAGGAACACCGGTGGCGAAGGCGGATCTCTGGGCCGATACTGACGCTGAGGAGCGAAAGCGTGGGGAGCGAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGGTGGGCACTAGGTGTGGGCGACATTCCACGTCGTCCGTGCCGCAGCTAACGCATTAAGTGCCCCGCCTGGGGAGTACGGCCGCAAGGCTAAAACTCAAAGGAATTGACGGGGGCCCGCACAAGCGGCGGAGCATGTGGCTTAATTCGACGCAACGCGAAGAACCTTACCAAGGCTTGACATACACCGGAAAGCATCAGAGATGGTGCCCCCCTTGTGGTCGGTGTACAGGTGGTGCATGGCTGTCGTCAGCTCGTGTCGTGAGATGTTGGGTTAAGTCCCGCAACGAGCGCAACCCTTGTCCCGTGTTGCCAGCAGGCCCTTGTGGTGCTGGGGACTCACGGGAGACCGCCGGGGTCAACTCGGAGGAAGGTGGGGACGACGTCAAGTCATCATGCCCCTTATGTCTTGGGCTGCACACGTGCTACAATGGCCGGTACAATGAGCTGCGATACCGCGAGGTGGAGCGAATCTCAAAAAGCCGGTCTCAGTTCGGATTGGGGTCTGCAACTCGACCCCATGAAGTCGGAGTCGCTAGTAATCGCAGATCAGCATTGCTGCGGTGAATACGTTCCCGGGCCTTGTACACACCGCCCGTCACGTCACGAAAGTCGGTAACACCCGAAGCCGGTGGCCCAACCCCTTGTGGGAGGGAGCTGTCGAAGGTGGGACTGGCGATTGGGACGAAGTCGTAACAAGGTAACCGATGCTCACCTAGAGCAATACGTAACTGAACAA'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'AGTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAACGGCAGCACAGGGGAGCTTGCTCCCTGGGTGGCGAGTGGCGGACGGGTGAGGAATACATCGGAATCTACCTTGTCGTGGGGGATAACGTAGGGAAACTTACGCTAATACCGCATACGACCTACGGGTGAAAGCGGAGGACCGCAAGGCTTCGCGCGATTGGATGAGCCGATGTCGGATTAGCTTGTTGGCGGGGTAACGGCCCACCAAGGCGACGATCCGTAGCTGGTCTGAGAGGATGATCAGCCACACTGGAACTGAGACACGGTCCAGACTCCTACGGGAGGCAGCAGTGGGGAATATTGGACAATGGGCGCAAGCCTGATCCAGCCATGCCGCGTGAGTGAAGAAGGCCCTCGGGTTGTAAAGCTCTTTTGTTGGGGAAGAAAAGCAATCGGTTAATACCCGGTTGTCCTGACGGTACCCAAAGAATAAGCACCGGCTAACTTCGTGCCAGCAGCCGCGGTAATACGAAGGGTGCAAGCGTTACTCGGAATTACTGGGCGTAAAGCGTGCGTAGGTGGTTTGTTAAGTCTGATGTGAAAGCCCTGGGCTCAACCTGGGAATTGCACTGGATACTGGCAGGCTAGAGTGCGGTAGAGGATGGCGGAATTCCCGGTGTAGCAGTGAAATGCGTAGAGATCGGGAGGAACATCTGTGGCGAAGGCGGCCATCTGGACCAGCACTGACACTGAGGCACGAAAGCGTGGGGAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCCTAAACGATGCGAACTGGATGTTGGGAGCAACTAGGCTCTCAGTATCGAAGCTAACGCGTTAAGTTCGCCGCCTGGGGAGTACGGTCGCAAGACTGAAACTCAAAGGAATTGACGGGGGCCCGCACAAGCGGTGGAGTATGTGGTTTAATTCGATGCAACGCGAAGAACCTTACCTGGCCTTGACATCCACGGAACTTACCAGAGATGGTTTGGTGCCTTCGGGAACCGTGAGACAGGTGCTGCATGGCTGTCGTCAGCTCGTGTCGTGAGATGTTGGGTTAAGTCCCGCAACGAGCGCAACCCTTGTCCTTAGTTGCCAGCACGTAATGGTGGGAACTCTAAGGAGACCGCCGGTGACAAACCGGAGGAAGGTGGGGATGACGTCAAGTCATCATGGCCCTTACGGCCAGGGCTACACACGTACTACAATGGAAAGGACAGAGGGCAGCAATCCCGCGAGGGGGAGCCAATCCCAGAAACCTTTTCTCAGTCCGGATCGGAGTCTGCAACTCGACTCCGTGAAGTCGGAATCGCTAGTAATCGCAGATCAGCATTGCTGCGGTGAATACGTTCCCGGGCCTTGTACACACCGCCCGTCACACCATGGGAGTGGGTTGCACCAGAAGTAGCTAGTCTAACCTTCGGGAGGACGGTTACCACGGTGTGATTCATGACTGGGGTGAAGTCGTAACAAGGTAACCGTGTCTACGCTAGAGCAATACGT'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'ATTGAACGCTGGCGGCAGGCCTGACACCGTGCAGATCAGACAATCAGAAGAGCTTGCTCTTCAATTCAGCGGCGGACGGGTGAGTAATGCCTAGGAATCTGCCTGGTAGTGGGGGACAACGTTTCGAAAGGAACGCTAATACCGCATACGTCCTACGGGAGAAAGCAGGGAGCCTCGGGCCTTGCGCTATCAGATGAGCCTAGGTCGGATTAGCTAGTTGGTGAGGTAATGGCTCACCAAGGCGACGATCCGTAACTGGTCTGAGAGGATGATCAGTCACACTGGAACTGAGACACGGTCCAGACTCCTACGGGAGGCAGCAGTGGGGAATCTGGACAATGGGCGAAAGCCTAGTCCAGCCATGCCGCGTGTGTGAAGAAGGTCTTCGGATTGTAAAGCACTTTAAGTTGGGAGGAGGGCATGCACCTAATACGTGTGTGTCTTGACGTTACCGACAGAATAAGCACCGGCTAACTCTGTGCCAGCAGCCGCGGTAATACAGAGGGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGGTGGTTCGTTGAGTTGAATGTGAAATCCCCGGGCTCAACCTGGGAACTGCATCAAAACTGGCAAGCTGAGTTCTGGTAGAGGGTGGTGGAATTTCCTGTGTAGCGGTGAGATGATAGGAACACCAGTGGCGAAGGCGACCACCTGGACTGATACTGACACTGAGGTGCGAGAGCGGGAAGCGAAGCGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGTCAACTAGCCGTTGGGAGCCTTGAGCTCTTAGTGGCGCAGCTAACGCATTAAGTTGACCGCACGCCTGGGGATGCGGCCGCAAGGTTAAAACTCAAATGAATTGACGGGGGCCCGCACAAGCGGTGGAGCATGTCGGATTAGATTGGAGGACAACGCGAAGAACCTTGCCTTGACATCAATGAACTTTCCAGAGATGGATTGGTGCCTTCGGGAACATTGAGACAGGTGCTGCATGGCTGTCGTCAGCTCGTGTCGTGAGATGTTGGGTTAAGTCCCGTAACGAGCGCAACCCTTGTCCTTAGTTACCAGCACGTTAAGGTGGGCACTACCGTCCGGTGACAAACCGGAGGAAGGTGGGGGATGGCTGGAATTAGTCATGGCCCTTACGGCCTACGCACGTGCTCAATCGTGGTACAAGAGGGTTGCCAAGCCGCGAGGTGGAGCTAATCCCAAGAAAAACCGATAATCCGGATCGCCGTCTGCAACTCGACTGCGTGAAGTCGGAATCGCTAGTAATCGCGAATCAGAATGTCGCGGTGAATACGTACCGGGCCTTGCACACACCACCGTCACACCAGTGGAATGGTGGCCCAGAAGTAGCTAGTCTAACCTTCGGGAGGACGGTTACCACGGTGTGATTCATGACTGGGGTGAAGTCGTAACAAGGTAGCCGGCGCATAGTCTAGAGCAATACG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>forward_primer</th>\n",
       "      <th>reverse_primer</th>\n",
       "      <th>descriptor</th>\n",
       "      <th>consensus_sequence</th>\n",
       "      <th>16S_seq_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>54ad1885-9646-4c70-8619-de9ae41c8821</td>\n",
       "      <td>TAAACCTCGTCGCGTGCTTAACACATGCAAGTCGAACGGTGAAGCC...</td>\n",
       "      <td>1469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>E</td>\n",
       "      <td>11</td>\n",
       "      <td>consensus_E_11_0_0(61)</td>\n",
       "      <td>GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGATGAA...</td>\n",
       "      <td>1500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>J</td>\n",
       "      <td>5</td>\n",
       "      <td>consensus_J_5_0_0(24)</td>\n",
       "      <td>AGTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAACGGCAGC...</td>\n",
       "      <td>1509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>J</td>\n",
       "      <td>18</td>\n",
       "      <td>0064a29b-3204-48b7-ae7b-8cfd0834c1d3</td>\n",
       "      <td>ATTGAACGCTGGCGGCAGGCCTGACACCGTGCAGATCAGACAATCA...</td>\n",
       "      <td>1468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>J</td>\n",
       "      <td>18</td>\n",
       "      <td>f963612b-fe12-4737-8f80-928f3ed57e0f</td>\n",
       "      <td>ATTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAGCGGATGA...</td>\n",
       "      <td>1501</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   forward_primer  reverse_primer                            descriptor  \\\n",
       "17              C               1  54ad1885-9646-4c70-8619-de9ae41c8821   \n",
       "32              E              11                consensus_E_11_0_0(61)   \n",
       "59              J               5                 consensus_J_5_0_0(24)   \n",
       "67              J              18  0064a29b-3204-48b7-ae7b-8cfd0834c1d3   \n",
       "68              J              18  f963612b-fe12-4737-8f80-928f3ed57e0f   \n",
       "\n",
       "                                   consensus_sequence  16S_seq_length  \n",
       "17  TAAACCTCGTCGCGTGCTTAACACATGCAAGTCGAACGGTGAAGCC...            1469  \n",
       "32  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGATGAA...            1500  \n",
       "59  AGTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAACGGCAGC...            1509  \n",
       "67  ATTGAACGCTGGCGGCAGGCCTGACACCGTGCAGATCAGACAATCA...            1468  \n",
       "68  ATTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAGCGGATGA...            1501  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df=pd.read_csv(\"class_16S_sequences_week_9.csv\")\n",
    "ct = pd.crosstab(df['reverse_primer'], df['forward_primer'])\n",
    "print(ct)\n",
    "consensus_sequence = df.query(\"(forward_primer == 'C' and reverse_primer == 1) | (forward_primer == 'J' and reverse_primer == 5)| (forward_primer == 'E' and reverse_primer == 11)| (forward_primer == 'J' and reverse_primer == 18)\")\n",
    "seq_list= consensus_sequence['consensus_sequence'].tolist()\n",
    "\n",
    "# Display both sequences if there are more than one\n",
    "if len(seq_list) > 0:\n",
    "    display(seq_list[0])  # First sequence\n",
    "if len(seq_list) > 1:\n",
    "    display(seq_list[1])  # Second sequence\n",
    "if len(seq_list) > 2:\n",
    "    display(seq_list[2]) \n",
    "if len(seq_list) > 3:\n",
    "    display(seq_list[3]) \n",
    "# Display the filtered consensus_sequence DataFrame\n",
    "display(consensus_sequence)\n",
    "\n",
    "#Your code here\n",
    "#c1=16 (seqlist0),e11=15(seqlist1), j5=c6(seqlist2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f918afb3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>GAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGTC...</td>\n",
       "      <td>Mycobacterium_heidelbergense(AJ000684)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bacteria;Proteobacteria;Gammaproteobacteria;Vi...</td>\n",
       "      <td>GTTTGATCCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...</td>\n",
       "      <td>Vibrio_atlanticus(EF599163)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>TAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAAC...</td>\n",
       "      <td>Mycobacterium_aubagnense(AY859683)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bacteria;Proteobacteria;Alphaproteobacteria;Rh...</td>\n",
       "      <td>AAGAGTTTGATCCTGGCTCAGAGCGAACGCTGGCGGCATGCTTAAC...</td>\n",
       "      <td>Acetobacter_senegalensis(AY883036)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bacteria;Proteobacteria;Betaproteobacteria;Bur...</td>\n",
       "      <td>ATTGAACGCTGGCGGCATGCCTTACACATGCAAGTCGAACGGTAAC...</td>\n",
       "      <td>Aquincola_tertiaricarbonis(DQ656489)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15150</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>AGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAACACA...</td>\n",
       "      <td>Leucobacter_komagatae(NR_112021.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15151</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>GATGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAG...</td>\n",
       "      <td>Thermobifida_fusca(NR_112015.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15152</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAA...</td>\n",
       "      <td>Microbispora_mesophila(NR_112014.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15153</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGAAAA...</td>\n",
       "      <td>Parvopolyspora_pallida(NR_112013.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15154</th>\n",
       "      <td>Bacteria;Proteobacteria;Gammaproteobacteria;En...</td>\n",
       "      <td>ACGCTGGCGGCAGGCCTAACACATGCAAGTCGAACGGTAGCACAGa...</td>\n",
       "      <td>Klebsiella_oxytoca(NR_112010.1)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15155 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      id  \\\n",
       "0      Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "1      Bacteria;Proteobacteria;Gammaproteobacteria;Vi...   \n",
       "2      Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "3      Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   \n",
       "4      Bacteria;Proteobacteria;Betaproteobacteria;Bur...   \n",
       "...                                                  ...   \n",
       "15150  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "15151  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "15152  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "15153  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "15154  Bacteria;Proteobacteria;Gammaproteobacteria;En...   \n",
       "\n",
       "                                                sequence  \\\n",
       "0      GAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGTC...   \n",
       "1      GTTTGATCCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...   \n",
       "2      TAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAAC...   \n",
       "3      AAGAGTTTGATCCTGGCTCAGAGCGAACGCTGGCGGCATGCTTAAC...   \n",
       "4      ATTGAACGCTGGCGGCATGCCTTACACATGCAAGTCGAACGGTAAC...   \n",
       "...                                                  ...   \n",
       "15150  AGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAACACA...   \n",
       "15151  GATGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAG...   \n",
       "15152  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAA...   \n",
       "15153  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGAAAA...   \n",
       "15154  ACGCTGGCGGCAGGCCTAACACATGCAAGTCGAACGGTAGCACAGa...   \n",
       "\n",
       "                                      species  \n",
       "0      Mycobacterium_heidelbergense(AJ000684)  \n",
       "1                 Vibrio_atlanticus(EF599163)  \n",
       "2          Mycobacterium_aubagnense(AY859683)  \n",
       "3          Acetobacter_senegalensis(AY883036)  \n",
       "4        Aquincola_tertiaricarbonis(DQ656489)  \n",
       "...                                       ...  \n",
       "15150      Leucobacter_komagatae(NR_112021.1)  \n",
       "15151         Thermobifida_fusca(NR_112015.1)  \n",
       "15152     Microbispora_mesophila(NR_112014.1)  \n",
       "15153     Parvopolyspora_pallida(NR_112013.1)  \n",
       "15154         Klebsiella_oxytoca(NR_112010.1)  \n",
       "\n",
       "[15155 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df1=pd.read_csv('16S_reference-Copy1.csv')\n",
    "display(df1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e6d6f10d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting levenshtein\n",
      "  Using cached levenshtein-0.27.1-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (160 kB)\n",
      "Collecting rapidfuzz<4.0.0,>=3.9.0\n",
      "  Downloading rapidfuzz-3.13.0-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m4.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hInstalling collected packages: rapidfuzz, levenshtein\n",
      "  Attempting uninstall: rapidfuzz\n",
      "    Found existing installation: rapidfuzz 3.2.0\n",
      "    Uninstalling rapidfuzz-3.2.0:\n",
      "      Successfully uninstalled rapidfuzz-3.2.0\n",
      "Successfully installed levenshtein-0.27.1 rapidfuzz-3.13.0\n"
     ]
    }
   ],
   "source": [
    "!pip install levenshtein\n",
    "#pip is a package manager:this command installs the levenshtein package on our Noteable instance."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a634c0aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>species</th>\n",
       "      <th>distance_0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>GAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGTC...</td>\n",
       "      <td>Mycobacterium_heidelbergense(AJ000684)</td>\n",
       "      <td>244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bacteria;Proteobacteria;Gammaproteobacteria;Vi...</td>\n",
       "      <td>GTTTGATCCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...</td>\n",
       "      <td>Vibrio_atlanticus(EF599163)</td>\n",
       "      <td>447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>TAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAAC...</td>\n",
       "      <td>Mycobacterium_aubagnense(AY859683)</td>\n",
       "      <td>239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bacteria;Proteobacteria;Alphaproteobacteria;Rh...</td>\n",
       "      <td>AAGAGTTTGATCCTGGCTCAGAGCGAACGCTGGCGGCATGCTTAAC...</td>\n",
       "      <td>Acetobacter_senegalensis(AY883036)</td>\n",
       "      <td>458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bacteria;Proteobacteria;Betaproteobacteria;Bur...</td>\n",
       "      <td>ATTGAACGCTGGCGGCATGCCTTACACATGCAAGTCGAACGGTAAC...</td>\n",
       "      <td>Aquincola_tertiaricarbonis(DQ656489)</td>\n",
       "      <td>420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGAAAG...</td>\n",
       "      <td>Couchioplanes_caeruleus(D85479)</td>\n",
       "      <td>219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Bacteria;Proteobacteria;Alphaproteobacteria;Rh...</td>\n",
       "      <td>ATGGCTCAGAACGAACGCTGGCGGCAGGCCTAACACATGCAAGTCG...</td>\n",
       "      <td>Maritimibacter_alkaliphilus(DQ915443)</td>\n",
       "      <td>433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>TCTGTAAGTAGCAGGAAAGGCCCTTCGGGGTACTCGAGCGGCGAAC...</td>\n",
       "      <td>Nocardioides_fonticola(EF626689)</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Bacteria;Proteobacteria;Gammaproteobacteria;Al...</td>\n",
       "      <td>AGGCCTAACACATGCAAGTCGAGCGGTAACACAAGGGAGCTTGCTC...</td>\n",
       "      <td>Shewanella_schlegeliana(AB081760)</td>\n",
       "      <td>423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAG...</td>\n",
       "      <td>Lentzea_albida(AB006176)</td>\n",
       "      <td>214</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  id  \\\n",
       "0  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "1  Bacteria;Proteobacteria;Gammaproteobacteria;Vi...   \n",
       "2  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "3  Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   \n",
       "4  Bacteria;Proteobacteria;Betaproteobacteria;Bur...   \n",
       "5  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "6  Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   \n",
       "7  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "8  Bacteria;Proteobacteria;Gammaproteobacteria;Al...   \n",
       "9  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  GAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGTC...   \n",
       "1  GTTTGATCCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...   \n",
       "2  TAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAAC...   \n",
       "3  AAGAGTTTGATCCTGGCTCAGAGCGAACGCTGGCGGCATGCTTAAC...   \n",
       "4  ATTGAACGCTGGCGGCATGCCTTACACATGCAAGTCGAACGGTAAC...   \n",
       "5  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGAAAG...   \n",
       "6  ATGGCTCAGAACGAACGCTGGCGGCAGGCCTAACACATGCAAGTCG...   \n",
       "7  TCTGTAAGTAGCAGGAAAGGCCCTTCGGGGTACTCGAGCGGCGAAC...   \n",
       "8  AGGCCTAACACATGCAAGTCGAGCGGTAACACAAGGGAGCTTGCTC...   \n",
       "9  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAG...   \n",
       "\n",
       "                                  species  distance_0  \n",
       "0  Mycobacterium_heidelbergense(AJ000684)         244  \n",
       "1             Vibrio_atlanticus(EF599163)         447  \n",
       "2      Mycobacterium_aubagnense(AY859683)         239  \n",
       "3      Acetobacter_senegalensis(AY883036)         458  \n",
       "4    Aquincola_tertiaricarbonis(DQ656489)         420  \n",
       "5         Couchioplanes_caeruleus(D85479)         219  \n",
       "6   Maritimibacter_alkaliphilus(DQ915443)         433  \n",
       "7        Nocardioides_fonticola(EF626689)         273  \n",
       "8       Shewanella_schlegeliana(AB081760)         423  \n",
       "9                Lentzea_albida(AB006176)         214  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>species</th>\n",
       "      <th>distance_0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8329</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>TGGCGGCGTGCTTAACACATGCAAGTCGAACGGTGAACACGGAGCT...</td>\n",
       "      <td>Microbacterium_oxydans(Y17227)</td>\n",
       "      <td>140</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     id  \\\n",
       "8329  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "\n",
       "                                               sequence  \\\n",
       "8329  TGGCGGCGTGCTTAACACATGCAAGTCGAACGGTGAACACGGAGCT...   \n",
       "\n",
       "                             species  distance_0  \n",
       "8329  Microbacterium_oxydans(Y17227)         140  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>species</th>\n",
       "      <th>distance_0</th>\n",
       "      <th>distance_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>GAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGTC...</td>\n",
       "      <td>Mycobacterium_heidelbergense(AJ000684)</td>\n",
       "      <td>244</td>\n",
       "      <td>228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bacteria;Proteobacteria;Gammaproteobacteria;Vi...</td>\n",
       "      <td>GTTTGATCCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...</td>\n",
       "      <td>Vibrio_atlanticus(EF599163)</td>\n",
       "      <td>447</td>\n",
       "      <td>411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>TAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAAC...</td>\n",
       "      <td>Mycobacterium_aubagnense(AY859683)</td>\n",
       "      <td>239</td>\n",
       "      <td>226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bacteria;Proteobacteria;Alphaproteobacteria;Rh...</td>\n",
       "      <td>AAGAGTTTGATCCTGGCTCAGAGCGAACGCTGGCGGCATGCTTAAC...</td>\n",
       "      <td>Acetobacter_senegalensis(AY883036)</td>\n",
       "      <td>458</td>\n",
       "      <td>435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bacteria;Proteobacteria;Betaproteobacteria;Bur...</td>\n",
       "      <td>ATTGAACGCTGGCGGCATGCCTTACACATGCAAGTCGAACGGTAAC...</td>\n",
       "      <td>Aquincola_tertiaricarbonis(DQ656489)</td>\n",
       "      <td>420</td>\n",
       "      <td>409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGAAAG...</td>\n",
       "      <td>Couchioplanes_caeruleus(D85479)</td>\n",
       "      <td>219</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Bacteria;Proteobacteria;Alphaproteobacteria;Rh...</td>\n",
       "      <td>ATGGCTCAGAACGAACGCTGGCGGCAGGCCTAACACATGCAAGTCG...</td>\n",
       "      <td>Maritimibacter_alkaliphilus(DQ915443)</td>\n",
       "      <td>433</td>\n",
       "      <td>400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>TCTGTAAGTAGCAGGAAAGGCCCTTCGGGGTACTCGAGCGGCGAAC...</td>\n",
       "      <td>Nocardioides_fonticola(EF626689)</td>\n",
       "      <td>273</td>\n",
       "      <td>256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Bacteria;Proteobacteria;Gammaproteobacteria;Al...</td>\n",
       "      <td>AGGCCTAACACATGCAAGTCGAGCGGTAACACAAGGGAGCTTGCTC...</td>\n",
       "      <td>Shewanella_schlegeliana(AB081760)</td>\n",
       "      <td>423</td>\n",
       "      <td>408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAG...</td>\n",
       "      <td>Lentzea_albida(AB006176)</td>\n",
       "      <td>214</td>\n",
       "      <td>163</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  id  \\\n",
       "0  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "1  Bacteria;Proteobacteria;Gammaproteobacteria;Vi...   \n",
       "2  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "3  Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   \n",
       "4  Bacteria;Proteobacteria;Betaproteobacteria;Bur...   \n",
       "5  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "6  Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   \n",
       "7  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "8  Bacteria;Proteobacteria;Gammaproteobacteria;Al...   \n",
       "9  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  GAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGTC...   \n",
       "1  GTTTGATCCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...   \n",
       "2  TAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAAC...   \n",
       "3  AAGAGTTTGATCCTGGCTCAGAGCGAACGCTGGCGGCATGCTTAAC...   \n",
       "4  ATTGAACGCTGGCGGCATGCCTTACACATGCAAGTCGAACGGTAAC...   \n",
       "5  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGAAAG...   \n",
       "6  ATGGCTCAGAACGAACGCTGGCGGCAGGCCTAACACATGCAAGTCG...   \n",
       "7  TCTGTAAGTAGCAGGAAAGGCCCTTCGGGGTACTCGAGCGGCGAAC...   \n",
       "8  AGGCCTAACACATGCAAGTCGAGCGGTAACACAAGGGAGCTTGCTC...   \n",
       "9  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAG...   \n",
       "\n",
       "                                  species  distance_0  distance_1  \n",
       "0  Mycobacterium_heidelbergense(AJ000684)         244         228  \n",
       "1             Vibrio_atlanticus(EF599163)         447         411  \n",
       "2      Mycobacterium_aubagnense(AY859683)         239         226  \n",
       "3      Acetobacter_senegalensis(AY883036)         458         435  \n",
       "4    Aquincola_tertiaricarbonis(DQ656489)         420         409  \n",
       "5         Couchioplanes_caeruleus(D85479)         219         182  \n",
       "6   Maritimibacter_alkaliphilus(DQ915443)         433         400  \n",
       "7        Nocardioides_fonticola(EF626689)         273         256  \n",
       "8       Shewanella_schlegeliana(AB081760)         423         408  \n",
       "9                Lentzea_albida(AB006176)         214         163  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>species</th>\n",
       "      <th>distance_0</th>\n",
       "      <th>distance_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6068</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>ACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGATGAAC...</td>\n",
       "      <td>Streptomyces_griseorubens(AB184139)</td>\n",
       "      <td>203</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     id  \\\n",
       "6068  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "\n",
       "                                               sequence  \\\n",
       "6068  ACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGATGAAC...   \n",
       "\n",
       "                                  species  distance_0  distance_1  \n",
       "6068  Streptomyces_griseorubens(AB184139)         203          39  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>species</th>\n",
       "      <th>distance_0</th>\n",
       "      <th>distance_1</th>\n",
       "      <th>distance_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>GAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGTC...</td>\n",
       "      <td>Mycobacterium_heidelbergense(AJ000684)</td>\n",
       "      <td>244</td>\n",
       "      <td>228</td>\n",
       "      <td>370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bacteria;Proteobacteria;Gammaproteobacteria;Vi...</td>\n",
       "      <td>GTTTGATCCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...</td>\n",
       "      <td>Vibrio_atlanticus(EF599163)</td>\n",
       "      <td>447</td>\n",
       "      <td>411</td>\n",
       "      <td>305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>TAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAAC...</td>\n",
       "      <td>Mycobacterium_aubagnense(AY859683)</td>\n",
       "      <td>239</td>\n",
       "      <td>226</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bacteria;Proteobacteria;Alphaproteobacteria;Rh...</td>\n",
       "      <td>AAGAGTTTGATCCTGGCTCAGAGCGAACGCTGGCGGCATGCTTAAC...</td>\n",
       "      <td>Acetobacter_senegalensis(AY883036)</td>\n",
       "      <td>458</td>\n",
       "      <td>435</td>\n",
       "      <td>408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bacteria;Proteobacteria;Betaproteobacteria;Bur...</td>\n",
       "      <td>ATTGAACGCTGGCGGCATGCCTTACACATGCAAGTCGAACGGTAAC...</td>\n",
       "      <td>Aquincola_tertiaricarbonis(DQ656489)</td>\n",
       "      <td>420</td>\n",
       "      <td>409</td>\n",
       "      <td>283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGAAAG...</td>\n",
       "      <td>Couchioplanes_caeruleus(D85479)</td>\n",
       "      <td>219</td>\n",
       "      <td>182</td>\n",
       "      <td>358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Bacteria;Proteobacteria;Alphaproteobacteria;Rh...</td>\n",
       "      <td>ATGGCTCAGAACGAACGCTGGCGGCAGGCCTAACACATGCAAGTCG...</td>\n",
       "      <td>Maritimibacter_alkaliphilus(DQ915443)</td>\n",
       "      <td>433</td>\n",
       "      <td>400</td>\n",
       "      <td>405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>TCTGTAAGTAGCAGGAAAGGCCCTTCGGGGTACTCGAGCGGCGAAC...</td>\n",
       "      <td>Nocardioides_fonticola(EF626689)</td>\n",
       "      <td>273</td>\n",
       "      <td>256</td>\n",
       "      <td>401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Bacteria;Proteobacteria;Gammaproteobacteria;Al...</td>\n",
       "      <td>AGGCCTAACACATGCAAGTCGAGCGGTAACACAAGGGAGCTTGCTC...</td>\n",
       "      <td>Shewanella_schlegeliana(AB081760)</td>\n",
       "      <td>423</td>\n",
       "      <td>408</td>\n",
       "      <td>292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAG...</td>\n",
       "      <td>Lentzea_albida(AB006176)</td>\n",
       "      <td>214</td>\n",
       "      <td>163</td>\n",
       "      <td>343</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  id  \\\n",
       "0  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "1  Bacteria;Proteobacteria;Gammaproteobacteria;Vi...   \n",
       "2  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "3  Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   \n",
       "4  Bacteria;Proteobacteria;Betaproteobacteria;Bur...   \n",
       "5  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "6  Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   \n",
       "7  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "8  Bacteria;Proteobacteria;Gammaproteobacteria;Al...   \n",
       "9  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  GAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGTC...   \n",
       "1  GTTTGATCCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...   \n",
       "2  TAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAAC...   \n",
       "3  AAGAGTTTGATCCTGGCTCAGAGCGAACGCTGGCGGCATGCTTAAC...   \n",
       "4  ATTGAACGCTGGCGGCATGCCTTACACATGCAAGTCGAACGGTAAC...   \n",
       "5  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGAAAG...   \n",
       "6  ATGGCTCAGAACGAACGCTGGCGGCAGGCCTAACACATGCAAGTCG...   \n",
       "7  TCTGTAAGTAGCAGGAAAGGCCCTTCGGGGTACTCGAGCGGCGAAC...   \n",
       "8  AGGCCTAACACATGCAAGTCGAGCGGTAACACAAGGGAGCTTGCTC...   \n",
       "9  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAG...   \n",
       "\n",
       "                                  species  distance_0  distance_1  distance_2  \n",
       "0  Mycobacterium_heidelbergense(AJ000684)         244         228         370  \n",
       "1             Vibrio_atlanticus(EF599163)         447         411         305  \n",
       "2      Mycobacterium_aubagnense(AY859683)         239         226         369  \n",
       "3      Acetobacter_senegalensis(AY883036)         458         435         408  \n",
       "4    Aquincola_tertiaricarbonis(DQ656489)         420         409         283  \n",
       "5         Couchioplanes_caeruleus(D85479)         219         182         358  \n",
       "6   Maritimibacter_alkaliphilus(DQ915443)         433         400         405  \n",
       "7        Nocardioides_fonticola(EF626689)         273         256         401  \n",
       "8       Shewanella_schlegeliana(AB081760)         423         408         292  \n",
       "9                Lentzea_albida(AB006176)         214         163         343  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>species</th>\n",
       "      <th>distance_0</th>\n",
       "      <th>distance_1</th>\n",
       "      <th>distance_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3661</th>\n",
       "      <td>Bacteria;Proteobacteria;Gammaproteobacteria;Xa...</td>\n",
       "      <td>GAGTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAACGGCAG...</td>\n",
       "      <td>Lysobacter_panaciterrae(AB245359)</td>\n",
       "      <td>394</td>\n",
       "      <td>376</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     id  \\\n",
       "3661  Bacteria;Proteobacteria;Gammaproteobacteria;Xa...   \n",
       "\n",
       "                                               sequence  \\\n",
       "3661  GAGTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAACGGCAG...   \n",
       "\n",
       "                                species  distance_0  distance_1  distance_2  \n",
       "3661  Lysobacter_panaciterrae(AB245359)         394         376          88  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from Levenshtein import distance\n",
    "\n",
    "# Compare to seq_list[0]\n",
    "string_to_compare_0 = seq_list[0]\n",
    "df1['distance_0'] = df1['sequence'].apply(distance, args=(string_to_compare_0,))\n",
    "\n",
    "# Display the updated DataFrame with distance_0\n",
    "display(df1.head(10))\n",
    "\n",
    "# Find the minimum distance for distance_0\n",
    "min_distance_0 = df1['distance_0'].min()\n",
    "best_match_0 = df1.query(\"distance_0 == @min_distance_0\")\n",
    "display(best_match_0)\n",
    "\n",
    "# Compare to seq_list[1]\n",
    "string_to_compare_1 = seq_list[1]\n",
    "df1['distance_1'] = df1['sequence'].apply(distance, args=(string_to_compare_1,))\n",
    "\n",
    "# Display the updated DataFrame with distance_1\n",
    "display(df1.head(10))\n",
    "\n",
    "# Find the minimum distance for distance_1\n",
    "min_distance_1 = df1['distance_1'].min()\n",
    "best_match_1 = df1.query(\"distance_1 == @min_distance_1\")\n",
    "display(best_match_1)\n",
    "\n",
    "# Compare to seq_list[2]\n",
    "string_to_compare_2 = seq_list[2]\n",
    "df1['distance_2'] = df1['sequence'].apply(distance, args=(string_to_compare_2,))\n",
    "\n",
    "# Display the updated DataFrame with distance_2\n",
    "display(df1.head(10))\n",
    "\n",
    "# Find the minimum distance for distance_2\n",
    "min_distance_2 = df1['distance_2'].min()\n",
    "best_match_2 = df1.query(\"distance_2 == @min_distance_2\")\n",
    "display(best_match_2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "929f4e4c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>species</th>\n",
       "      <th>distance_0</th>\n",
       "      <th>distance_1</th>\n",
       "      <th>distance_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>GAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGTC...</td>\n",
       "      <td>Mycobacterium_heidelbergense(AJ000684)</td>\n",
       "      <td>244</td>\n",
       "      <td>228</td>\n",
       "      <td>370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bacteria;Proteobacteria;Gammaproteobacteria;Vi...</td>\n",
       "      <td>GTTTGATCCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...</td>\n",
       "      <td>Vibrio_atlanticus(EF599163)</td>\n",
       "      <td>447</td>\n",
       "      <td>411</td>\n",
       "      <td>305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>TAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAAC...</td>\n",
       "      <td>Mycobacterium_aubagnense(AY859683)</td>\n",
       "      <td>239</td>\n",
       "      <td>226</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bacteria;Proteobacteria;Alphaproteobacteria;Rh...</td>\n",
       "      <td>AAGAGTTTGATCCTGGCTCAGAGCGAACGCTGGCGGCATGCTTAAC...</td>\n",
       "      <td>Acetobacter_senegalensis(AY883036)</td>\n",
       "      <td>458</td>\n",
       "      <td>435</td>\n",
       "      <td>408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bacteria;Proteobacteria;Betaproteobacteria;Bur...</td>\n",
       "      <td>ATTGAACGCTGGCGGCATGCCTTACACATGCAAGTCGAACGGTAAC...</td>\n",
       "      <td>Aquincola_tertiaricarbonis(DQ656489)</td>\n",
       "      <td>420</td>\n",
       "      <td>409</td>\n",
       "      <td>283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGAAAG...</td>\n",
       "      <td>Couchioplanes_caeruleus(D85479)</td>\n",
       "      <td>219</td>\n",
       "      <td>182</td>\n",
       "      <td>358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Bacteria;Proteobacteria;Alphaproteobacteria;Rh...</td>\n",
       "      <td>ATGGCTCAGAACGAACGCTGGCGGCAGGCCTAACACATGCAAGTCG...</td>\n",
       "      <td>Maritimibacter_alkaliphilus(DQ915443)</td>\n",
       "      <td>433</td>\n",
       "      <td>400</td>\n",
       "      <td>405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>TCTGTAAGTAGCAGGAAAGGCCCTTCGGGGTACTCGAGCGGCGAAC...</td>\n",
       "      <td>Nocardioides_fonticola(EF626689)</td>\n",
       "      <td>273</td>\n",
       "      <td>256</td>\n",
       "      <td>401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Bacteria;Proteobacteria;Gammaproteobacteria;Al...</td>\n",
       "      <td>AGGCCTAACACATGCAAGTCGAGCGGTAACACAAGGGAGCTTGCTC...</td>\n",
       "      <td>Shewanella_schlegeliana(AB081760)</td>\n",
       "      <td>423</td>\n",
       "      <td>408</td>\n",
       "      <td>292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAG...</td>\n",
       "      <td>Lentzea_albida(AB006176)</td>\n",
       "      <td>214</td>\n",
       "      <td>163</td>\n",
       "      <td>343</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  id  \\\n",
       "0  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "1  Bacteria;Proteobacteria;Gammaproteobacteria;Vi...   \n",
       "2  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "3  Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   \n",
       "4  Bacteria;Proteobacteria;Betaproteobacteria;Bur...   \n",
       "5  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "6  Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   \n",
       "7  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "8  Bacteria;Proteobacteria;Gammaproteobacteria;Al...   \n",
       "9  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  GAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGTC...   \n",
       "1  GTTTGATCCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...   \n",
       "2  TAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAAC...   \n",
       "3  AAGAGTTTGATCCTGGCTCAGAGCGAACGCTGGCGGCATGCTTAAC...   \n",
       "4  ATTGAACGCTGGCGGCATGCCTTACACATGCAAGTCGAACGGTAAC...   \n",
       "5  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGAAAG...   \n",
       "6  ATGGCTCAGAACGAACGCTGGCGGCAGGCCTAACACATGCAAGTCG...   \n",
       "7  TCTGTAAGTAGCAGGAAAGGCCCTTCGGGGTACTCGAGCGGCGAAC...   \n",
       "8  AGGCCTAACACATGCAAGTCGAGCGGTAACACAAGGGAGCTTGCTC...   \n",
       "9  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAG...   \n",
       "\n",
       "                                  species  distance_0  distance_1  distance_2  \n",
       "0  Mycobacterium_heidelbergense(AJ000684)         244         228         370  \n",
       "1             Vibrio_atlanticus(EF599163)         447         411         305  \n",
       "2      Mycobacterium_aubagnense(AY859683)         239         226         369  \n",
       "3      Acetobacter_senegalensis(AY883036)         458         435         408  \n",
       "4    Aquincola_tertiaricarbonis(DQ656489)         420         409         283  \n",
       "5         Couchioplanes_caeruleus(D85479)         219         182         358  \n",
       "6   Maritimibacter_alkaliphilus(DQ915443)         433         400         405  \n",
       "7        Nocardioides_fonticola(EF626689)         273         256         401  \n",
       "8       Shewanella_schlegeliana(AB081760)         423         408         292  \n",
       "9                Lentzea_albida(AB006176)         214         163         343  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>species</th>\n",
       "      <th>distance_0</th>\n",
       "      <th>distance_1</th>\n",
       "      <th>distance_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8329</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>TGGCGGCGTGCTTAACACATGCAAGTCGAACGGTGAACACGGAGCT...</td>\n",
       "      <td>Microbacterium_oxydans(Y17227)</td>\n",
       "      <td>140</td>\n",
       "      <td>202</td>\n",
       "      <td>349</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     id  \\\n",
       "8329  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "\n",
       "                                               sequence  \\\n",
       "8329  TGGCGGCGTGCTTAACACATGCAAGTCGAACGGTGAACACGGAGCT...   \n",
       "\n",
       "                             species  distance_0  distance_1  distance_2  \n",
       "8329  Microbacterium_oxydans(Y17227)         140         202         349  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>species</th>\n",
       "      <th>distance_0</th>\n",
       "      <th>distance_1</th>\n",
       "      <th>distance_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6068</th>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>ACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGATGAAC...</td>\n",
       "      <td>Streptomyces_griseorubens(AB184139)</td>\n",
       "      <td>203</td>\n",
       "      <td>39</td>\n",
       "      <td>347</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     id  \\\n",
       "6068  Bacteria;Actinobacteria;Actinobacteria;Actinom...   \n",
       "\n",
       "                                               sequence  \\\n",
       "6068  ACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGATGAAC...   \n",
       "\n",
       "                                  species  distance_0  distance_1  distance_2  \n",
       "6068  Streptomyces_griseorubens(AB184139)         203          39         347  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>species</th>\n",
       "      <th>distance_0</th>\n",
       "      <th>distance_1</th>\n",
       "      <th>distance_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3661</th>\n",
       "      <td>Bacteria;Proteobacteria;Gammaproteobacteria;Xa...</td>\n",
       "      <td>GAGTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAACGGCAG...</td>\n",
       "      <td>Lysobacter_panaciterrae(AB245359)</td>\n",
       "      <td>394</td>\n",
       "      <td>376</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     id  \\\n",
       "3661  Bacteria;Proteobacteria;Gammaproteobacteria;Xa...   \n",
       "\n",
       "                                               sequence  \\\n",
       "3661  GAGTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAACGGCAG...   \n",
       "\n",
       "                                species  distance_0  distance_1  distance_2  \n",
       "3661  Lysobacter_panaciterrae(AB245359)         394         376          88  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# First, check if there's an existing 'distance' column and drop it\n",
    "if 'distance' in df1.columns:\n",
    "    df1.drop(columns=['distance'], inplace=True)\n",
    "\n",
    "# Now calculate the Levenshtein distance for each sequence\n",
    "string_to_compare_0 = seq_list[0]\n",
    "df1['distance_0'] = df1['sequence'].apply(distance, args=(string_to_compare_0,))\n",
    "\n",
    "string_to_compare_1 = seq_list[1]\n",
    "df1['distance_1'] = df1['sequence'].apply(distance, args=(string_to_compare_1,))\n",
    "\n",
    "string_to_compare_2 = seq_list[2]\n",
    "df1['distance_2'] = df1['sequence'].apply(distance, args=(string_to_compare_2,))\n",
    "\n",
    "# Display the DataFrame with the updated distance columns\n",
    "display(df1.head(10))\n",
    "\n",
    "# Calculate the minimum distances and find the best matches\n",
    "min_distance_0 = df1['distance_0'].min()\n",
    "best_match_0 = df1.query(\"distance_0 == @min_distance_0\")\n",
    "display(best_match_0)\n",
    "\n",
    "min_distance_1 = df1['distance_1'].min()\n",
    "best_match_1 = df1.query(\"distance_1 == @min_distance_1\")\n",
    "display(best_match_1)\n",
    "\n",
    "min_distance_2 = df1['distance_2'].min()\n",
    "best_match_2 = df1.query(\"distance_2 == @min_distance_2\")\n",
    "display(best_match_2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c4382a5",
   "metadata": {},
   "source": [
    "## Commentary\n",
    "\n",
    "Three colonies—colony 15, colony 16 and colony C6—were selected for sequencing since they show inhibition against pathogens in overlay test. For colony 15, the closest species match is Streptomyces griseorubens (AB184139), with Levenshtein distance of 39. The closest match to colony 16 is Microbacterium oxydans (Y17227), having a distance of 140. Lastly, Lysobacter panaciterrae (AB245329) is the closest match to colony C6, with a Levenshtein distance of 88.\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
